{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Refilter all. Aligning to genome makes no missing data. Adjust to remove loci private in each ssp/ploidy group too "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/usr/bin/sh: line 0: source: activate: file not found\r\n"
     ]
    }
   ],
   "source": [
    "!source activate py36"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys\n",
    "\n",
    "#sys.path.append('/home/faske/g/anaconda3/envs/py34/lib/python3.4/site-packages')\n",
    "sys.path.append('/data/gpfs/assoc/parchmanlab/tfaske/anaconda3/envs/py36/lib/python3.6/site-packages')\n",
    "sys.path.append(\"/data/gpfs/assoc/parchmanlab/tfaske/ipynb/include_utils\")\n",
    "\n",
    "import ipyparallel as ipp\n",
    "import os, time\n",
    "import include_utils as u\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import scipy as sp\n",
    "import random\n",
    "import numbers\n",
    "import matplotlib.pyplot as plt\n",
    "import matplotlib.patches as mpatches\n",
    "import matplotlib.cm as cm\n",
    "import matplotlib.colors as mcolors\n",
    "#import vcf\n",
    "from sklearn import preprocessing\n",
    "from subprocess import Popen, PIPE, call, check_output\n",
    "import seaborn as sns\n",
    "from IPython.display import FileLink\n",
    "import urllib.request as urllib2\n",
    "import dill\n",
    "import traceback\n",
    "from pandas import Series, DataFrame\n",
    "import gzip\n",
    "import warnings\n",
    "warnings.filterwarnings('ignore',category=pd.io.pytables.PerformanceWarning)\n",
    "%config InlineBackend.figure_format = 'retina'\n",
    "from Bio import SeqIO\n",
    "#import pysam\n",
    "from collections import OrderedDict, namedtuple, Counter\n",
    "import operator\n",
    "import multiprocessing as mp\n",
    "import shutil\n",
    "import tempfile\n",
    "#from ipyparallel import Client\n",
    "import scandir\n",
    "import glob\n",
    "from Bio.SeqIO.QualityIO import FastqGeneralIterator\n",
    "import pickle\n",
    "import re\n",
    "from itertools import chain\n",
    "#import Levenshtein as lv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "vcfutils = \"vcfutils.pl\"\n",
    "vcftools = \"vcftools\"\n",
    "bcftools = \"bcftools\"\n",
    "bgzip = \"bgzip\"\n",
    "tabix = \"tabix\"\n",
    "java = \"java\"\n",
    "plink = \"plink --noweb\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "analysis_dir = \"/data/gpfs/assoc/denovo/tfaske/sagebrush/SNPcall/filtering\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/data/gpfs/assoc/denovo/tfaske/sagebrush/SNPcall/filtering\n"
     ]
    }
   ],
   "source": [
    "cd $analysis_dir"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'/data/gpfs/assoc/denovo/tfaske/sagebrush/SNPcall/filtering/ARTR_chr.vcf.gz'"
      ]
     },
     "execution_count": 73,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "vcf_file = os.path.join(analysis_dir, \"ARTR_chr.vcf.gz\")\n",
    "assert os.path.exists(vcf_file)\n",
    "vcf_file"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## keep only biallelic first"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "VCFtools - 0.1.16\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf /data/gpfs/assoc/denovo/tfaske/sagebrush/SNPcall/filtering/ARTR_chr.vcf.gz\n",
      "\t--recode-INFO-all\n",
      "\t--max-alleles 2\n",
      "\t--min-alleles 2\n",
      "\t--out ARTR_chr.biallelic\n",
      "\t--recode\n",
      "\t--remove-filtered-all\n",
      "\t--remove-indels\n",
      "\n",
      "Using zlib version: 1.2.11\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=PV4,Number=4,Type=Float,Description=\"P-values for strand bias, baseQ bias, mapQ bias and tail distance bias\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=PV4,Number=4,Type=Float,Description=\"P-values for strand bias, baseQ bias, mapQ bias and tail distance bias\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=DP4,Number=4,Type=Integer,Description=\"Number of high-quality ref-forward , ref-reverse, alt-forward and alt-reverse bases\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=DP4,Number=4,Type=Integer,Description=\"Number of high-quality ref-forward , ref-reverse, alt-forward and alt-reverse bases\">\n",
      "After filtering, kept 708 out of 708 Individuals\n",
      "Outputting VCF file...\n",
      "After filtering, kept 2766918 out of a possible 2774609 Sites\n",
      "Run Time = 3749.00 seconds\n"
     ]
    }
   ],
   "source": [
    "!$vcftools --remove-indels \\\n",
    "--min-alleles 2 \\\n",
    "--max-alleles 2 \\\n",
    "--remove-filtered-all \\\n",
    "--recode \\\n",
    "--recode-INFO-all \\\n",
    "--gzvcf \\\n",
    "$vcf_file \\\n",
    "--out $'ARTR_chr.biallelic'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 223,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'/data/gpfs/assoc/denovo/tfaske/sagebrush/SNPcall/filtering/ARTR_chr.biallelic.recode.vcf.gz'"
      ]
     },
     "execution_count": 223,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#already done\n",
    "#vcf_biallelic = \"ARTR_chr.biallelic.recode.vcf\" \n",
    "#vcf_biallelic_gz = \"%s.gz\" % vcf_biallelic\n",
    "#!$bgzip -c $vcf_biallelic > {vcf_biallelic_gz}\n",
    "#!$tabix {vcf_biallelic_gz}\n",
    "\n",
    "vcf_biallelic_gz = os.path.join(analysis_dir, \"ARTR_chr.biallelic.recode.vcf.gz\")\n",
    "assert os.path.exists(vcf_file)\n",
    "vcf_biallelic_gz"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Remove by MAF, missing, and thin next"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 225,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "VCFtools - 0.1.16\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf /data/gpfs/assoc/denovo/tfaske/sagebrush/SNPcall/filtering/ARTR_chr.biallelic.recode.vcf.gz\n",
      "\t--recode-INFO-all\n",
      "\t--maf 0.01\n",
      "\t--thin 100\n",
      "\t--max-missing 0.7\n",
      "\t--out ARTR_chr_miss70_thin100_MAF1\n",
      "\t--recode\n",
      "\t--remove-filtered-all\n",
      "\n",
      "Using zlib version: 1.2.11\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=PV4,Number=4,Type=Float,Description=\"P-values for strand bias, baseQ bias, mapQ bias and tail distance bias\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=PV4,Number=4,Type=Float,Description=\"P-values for strand bias, baseQ bias, mapQ bias and tail distance bias\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=DP4,Number=4,Type=Integer,Description=\"Number of high-quality ref-forward , ref-reverse, alt-forward and alt-reverse bases\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=DP4,Number=4,Type=Integer,Description=\"Number of high-quality ref-forward , ref-reverse, alt-forward and alt-reverse bases\">\n",
      "After filtering, kept 708 out of 708 Individuals\n",
      "Outputting VCF file...\n",
      "After filtering, kept 470121 out of a possible 2766918 Sites\n",
      "Run Time = 878.00 seconds\n"
     ]
    }
   ],
   "source": [
    "!$vcftools \\\n",
    "--max-missing 0.7 \\\n",
    "--maf 0.01 \\\n",
    "--thin 100 \\\n",
    "--remove-filtered-all \\\n",
    "--recode \\\n",
    "--recode-INFO-all \\\n",
    "--gzvcf \\\n",
    "$vcf_biallelic_gz \\\n",
    "--out $'ARTR_chr_miss70_thin100_MAF1'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [],
   "source": [
    "vcf_filtered = \"ARTR_chr_miss70_thin100_MAF1.recode.vcf\"\n",
    "vcf_filtered_gz = \"%s.gz\" % vcf_filtered"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 227,
   "metadata": {},
   "outputs": [],
   "source": [
    "!$bgzip -c $vcf_filtered > {vcf_filtered_gz}\n",
    "!$tabix {vcf_filtered_gz}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Calulate missing data and private alleles using vcfR\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "%load_ext rpy2.ipython"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "R[write to console]: Loading required package: readr\n",
      "\n",
      "R[write to console]: Loading required package: MASS\n",
      "\n",
      "R[write to console]: Loading required package: ggplot2\n",
      "\n",
      "R[write to console]: \n",
      "   *****       ***   vcfR   ***       *****\n",
      "   This is vcfR 1.8.0 \n",
      "     browseVignettes('vcfR') # Documentation\n",
      "     citation('vcfR') # Citation\n",
      "   *****       *****      *****       *****\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "%%R\n",
    "require(readr)\n",
    "library(data.table)\n",
    "require(MASS)\n",
    "require(ggplot2)\n",
    "library(vcfR)\n",
    "\n",
    "source('/data/gpfs/home/tfaske/g/src/R/Imports.R')\n",
    "\n",
    "setwd('/data/gpfs/assoc/denovo/tfaske/sagebrush/SNPcall/filtering/')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R\n",
    "vcf <- read.vcfR(\"ARTR_chr_miss70_thin100_MAF1.recode.vcf\", verbose = FALSE)\n",
    "\n",
    "#get positions\n",
    "chrom <- getCHROM(vcf)\n",
    "pos <- getPOS(vcf)\n",
    "pos_ID <- paste(chrom,pos,sep = ':')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R\n",
    "#get pl \n",
    "dp <- extract.gt(vcf, element = 'DP')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] 470121\n",
      " [1] \"CM042338.1:22493\"  \"CM042338.1:59115\"  \"CM042338.1:97078\" \n",
      " [4] \"CM042338.1:106063\" \"CM042338.1:126072\" \"CM042338.1:128439\"\n",
      " [7] \"CM042338.1:129199\" \"CM042338.1:129763\" \"CM042338.1:130233\"\n",
      "[10] \"CM042338.1:169173\"\n",
      "                  AT2_BB_10 AT2_BB_11 AT2_BB_12 AT2_BB_13 AT2_BB_14\n",
      "CM042338.1_22493  \"0\"       \"0\"       \"0\"       \"0\"       \"0\"      \n",
      "CM042338.1_59115  \"3\"       \"0\"       \"0\"       \"0\"       \"0\"      \n",
      "CM042338.1_97078  \"0\"       \"0\"       \"0\"       \"0\"       \"0\"      \n",
      "CM042338.1_106063 \"0\"       \"0\"       \"0\"       \"0\"       \"0\"      \n",
      "CM042338.1_126072 \"2\"       \"0\"       \"2\"       \"1\"       \"4\"      \n"
     ]
    }
   ],
   "source": [
    "%%R\n",
    "## check out PL and pos_ID\n",
    "print(length(pos_ID))\n",
    "print(pos_ID[1:10])\n",
    "dp[1:5,1:5]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " chr [1:5, 1:5] \"0\" \"3\" \"0\" \"0\" \"2\" \"0\" \"0\" \"0\" \"0\" \"0\" \"0\" \"0\" \"0\" \"0\" \"2\" ...\n",
      " - attr(*, \"dimnames\")=List of 2\n",
      "  ..$ : chr [1:5] \"CM042338.1_22493\" \"CM042338.1_59115\" \"CM042338.1_97078\" \"CM042338.1_106063\" ...\n",
      "  ..$ : chr [1:5] \"AT2_BB_10\" \"AT2_BB_11\" \"AT2_BB_12\" \"AT2_BB_13\" ...\n"
     ]
    }
   ],
   "source": [
    "%%R\n",
    "str(dp[1:5,1:5])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R\n",
    "pl_sum <- apply(pl,1:2, function(GP) sum(as.numeric(unlist(strsplit(as.character(GP),split = ',')))))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Calculate missing loci and individual"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] 708\n",
      "[1] 470121\n"
     ]
    }
   ],
   "source": [
    "%%R\n",
    "nloci <- ncol(dp)\n",
    "nindv <- nrow(dp)\n",
    "print(nloci)\n",
    "print(nindv)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "    Min.  1st Qu.   Median     Mean  3rd Qu.     Max. \n",
      "0.001412 0.980226 0.998588 0.908238 0.998588 0.998588 \n"
     ]
    }
   ],
   "source": [
    "%%R\n",
    "miss_loci <- apply(dp,1, function(d) length(which(d == \"0\"))/nloci)\n",
    "#miss_loci <- apply(dp[,1:100],1, function(d) (length(which(d == \"0\"))/100))\n",
    "print(summary(miss_loci))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] 15690\n",
      "[1] 24759\n",
      "[1] 35149\n",
      "[1] 46854\n",
      "[1] 60896\n"
     ]
    }
   ],
   "source": [
    "%%R\n",
    "print(length(which(miss_loci <= .30)))\n",
    "print(length(which(miss_loci <= .40)))\n",
    "print(length(which(miss_loci <= .50)))\n",
    "print(length(which(miss_loci <= .60)))\n",
    "print(length(which(miss_loci <= .70)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R\n",
    "keep_miss <- pos_ID[which(miss_loci <= .60)]\n",
    "length(keep_miss)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                        chrom     pos\n",
      "CM042338.1:126072  CM042338.1  126072\n",
      "CM042338.1:1342981 CM042338.1 1342981\n",
      "CM042338.1:1496841 CM042338.1 1496841\n",
      "CM042338.1:1744262 CM042338.1 1744262\n",
      "CM042338.1:1942811 CM042338.1 1942811\n",
      "CM042338.1:2385838 CM042338.1 2385838\n"
     ]
    }
   ],
   "source": [
    "%%R\n",
    "keep_miss60_df <- data.frame(chrom=sapply(keep_miss,function(s) unlist(strsplit(as.character(s),':'))[1]),\n",
    "                             pos=sapply(keep_miss,function(s) unlist(strsplit(as.character(s),':'))[2]))\n",
    "print(head(keep_miss60_df))    \n",
    "write.table(keep_miss60_df,'keep_miss60.txt',quote=F,row.names=F,col.names=F,sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CM042338.1\t126072\r\n",
      "CM042338.1\t1342981\r\n",
      "CM042338.1\t1496841\r\n",
      "CM042338.1\t1744262\r\n",
      "CM042338.1\t1942811\r\n",
      "CM042338.1\t2385838\r\n",
      "CM042338.1\t3792582\r\n",
      "CM042338.1\t4255378\r\n",
      "CM042338.1\t5373930\r\n",
      "CM042338.1\t6087026\r\n"
     ]
    }
   ],
   "source": [
    "!head keep_miss60.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "VCFtools - 0.1.16\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf ARTR_chr_miss70_thin100_MAF1.recode.vcf.gz\n",
      "\t--recode-INFO-all\n",
      "\t--out /data/gpfs/assoc/denovo/tfaske/sagebrush/SNPcall/filtering/ARTR_chr_miss60_thin100_MAF1\n",
      "\t--positions /data/gpfs/assoc/denovo/tfaske/sagebrush/SNPcall/filtering/keep_miss60.txt\n",
      "\t--recode\n",
      "\t--remove-filtered-all\n",
      "\n",
      "Using zlib version: 1.2.11\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=PV4,Number=4,Type=Float,Description=\"P-values for strand bias, baseQ bias, mapQ bias and tail distance bias\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=PV4,Number=4,Type=Float,Description=\"P-values for strand bias, baseQ bias, mapQ bias and tail distance bias\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=DP4,Number=4,Type=Integer,Description=\"Number of high-quality ref-forward , ref-reverse, alt-forward and alt-reverse bases\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=DP4,Number=4,Type=Integer,Description=\"Number of high-quality ref-forward , ref-reverse, alt-forward and alt-reverse bases\">\n",
      "After filtering, kept 708 out of 708 Individuals\n",
      "Outputting VCF file...\n",
      "After filtering, kept 46854 out of a possible 470121 Sites\n",
      "Run Time = 76.00 seconds\n"
     ]
    }
   ],
   "source": [
    "!$vcftools --gzvcf $vcf_filtered_gz \\\n",
    "--remove-filtered-all \\\n",
    "--recode \\\n",
    "--recode-INFO-all \\\n",
    "--positions {os.path.join(analysis_dir, \"keep_miss60.txt\")} \\\n",
    "--out {os.path.join(analysis_dir, \"ARTR_chr_miss60_thin100_MAF1\")}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [],
   "source": [
    "vcf_filtered = \"ARTR_chr_miss60_thin100_MAF1.recode.vcf\"\n",
    "vcf_filtered_gz = \"%s.gz\" % vcf_filtered\n",
    "!$bgzip -c $vcf_filtered > {vcf_filtered_gz}\n",
    "!$tabix {vcf_filtered_gz}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Remove bad_indv\n",
    "\n",
    "by do it by mean_depth cause no missing data "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "VCFtools - 0.1.16\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf ARTR_chr_miss60_thin100_MAF1.recode.vcf.gz\n",
      "\t--depth\n",
      "\t--out ARTR_chr_miss60_thin100_MAF1.recode.vcf.gz\n",
      "\n",
      "Using zlib version: 1.2.11\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=PV4,Number=4,Type=Float,Description=\"P-values for strand bias, baseQ bias, mapQ bias and tail distance bias\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=PV4,Number=4,Type=Float,Description=\"P-values for strand bias, baseQ bias, mapQ bias and tail distance bias\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=DP4,Number=4,Type=Integer,Description=\"Number of high-quality ref-forward , ref-reverse, alt-forward and alt-reverse bases\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=DP4,Number=4,Type=Integer,Description=\"Number of high-quality ref-forward , ref-reverse, alt-forward and alt-reverse bases\">\n",
      "After filtering, kept 708 out of 708 Individuals\n",
      "Outputting Mean Depth by Individual\n",
      "After filtering, kept 46854 out of a possible 46854 Sites\n",
      "Run Time = 6.00 seconds\n"
     ]
    }
   ],
   "source": [
    "!$vcftools --gzvcf {vcf_filtered_gz} \\\n",
    "--out {vcf_filtered_gz} \\\n",
    "--depth"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>INDV</th>\n",
       "      <th>N_SITES</th>\n",
       "      <th>MEAN_DEPTH</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AT2_BB_10</td>\n",
       "      <td>46854</td>\n",
       "      <td>2.67627</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AT2_BB_11</td>\n",
       "      <td>46854</td>\n",
       "      <td>1.75917</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AT2_BB_12</td>\n",
       "      <td>46854</td>\n",
       "      <td>1.50275</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AT2_BB_13</td>\n",
       "      <td>46854</td>\n",
       "      <td>2.21315</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AT2_BB_14</td>\n",
       "      <td>46854</td>\n",
       "      <td>3.08396</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        INDV  N_SITES  MEAN_DEPTH\n",
       "0  AT2_BB_10    46854     2.67627\n",
       "1  AT2_BB_11    46854     1.75917\n",
       "2  AT2_BB_12    46854     1.50275\n",
       "3  AT2_BB_13    46854     2.21315\n",
       "4  AT2_BB_14    46854     3.08396"
      ]
     },
     "execution_count": 41,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "depth_file = vcf_filtered_gz + \".idepth\"\n",
    "depth_df = pd.read_csv(depth_file, sep=\"\\t\")\n",
    "depth_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "count    708.000000\n",
       "mean       2.050820\n",
       "std        0.758901\n",
       "min        0.004781\n",
       "25%        1.546605\n",
       "50%        1.983460\n",
       "75%        2.606032\n",
       "max        4.326910\n",
       "Name: MEAN_DEPTH, dtype: float64"
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "depth_df.MEAN_DEPTH.describe()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(708, 6, 10, 12)"
      ]
     },
     "execution_count": 43,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(depth_df),len(depth_df[depth_df.MEAN_DEPTH <= .05]),len(depth_df[depth_df.MEAN_DEPTH <= .1]), len(depth_df[depth_df.MEAN_DEPTH <= .15])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [],
   "source": [
    "bad_indv = depth_df.INDV[depth_df.MEAN_DEPTH <= .1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(os.path.join(analysis_dir, \"bad_indv.txt\"), \"w\") as o:\n",
    "    o.write(\"INDV\\n\")\n",
    "    for elem in bad_indv:\n",
    "        o.write(\"%s\\n\" % elem)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "INDV\r\n",
      "AT2_LN_97\r\n",
      "AT2_OC_1\r\n",
      "AT2_SS_9\r\n",
      "AT4_DS_9\r\n",
      "AV2_TF_149\r\n",
      "AV4_EW_5\r\n",
      "AW4_FR_4\r\n",
      "AW4_GB_3\r\n",
      "AW4_JC_2\r\n",
      "AW4_MK_232\r\n"
     ]
    }
   ],
   "source": [
    "!cat bad_indv.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "VCFtools - 0.1.16\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf ARTR_chr_miss60_thin100_MAF1.recode.vcf.gz\n",
      "\t--remove /data/gpfs/assoc/denovo/tfaske/sagebrush/SNPcall/filtering/bad_indv.txt\n",
      "\t--recode-INFO-all\n",
      "\t--out /data/gpfs/assoc/denovo/tfaske/sagebrush/SNPcall/filtering/snps_indv_removed\n",
      "\t--recode\n",
      "\t--remove-filtered-all\n",
      "\t--remove-indels\n",
      "\n",
      "Using zlib version: 1.2.11\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=PV4,Number=4,Type=Float,Description=\"P-values for strand bias, baseQ bias, mapQ bias and tail distance bias\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=PV4,Number=4,Type=Float,Description=\"P-values for strand bias, baseQ bias, mapQ bias and tail distance bias\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=DP4,Number=4,Type=Integer,Description=\"Number of high-quality ref-forward , ref-reverse, alt-forward and alt-reverse bases\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=DP4,Number=4,Type=Integer,Description=\"Number of high-quality ref-forward , ref-reverse, alt-forward and alt-reverse bases\">\n",
      "Excluding individuals in 'exclude' list\n",
      "After filtering, kept 698 out of 708 Individuals\n",
      "Outputting VCF file...\n",
      "After filtering, kept 46854 out of a possible 46854 Sites\n",
      "Run Time = 65.00 seconds\n"
     ]
    }
   ],
   "source": [
    "!$vcftools --gzvcf $vcf_filtered_gz \\\n",
    "--remove-indels  \\\n",
    "--remove-filtered-all \\\n",
    "--recode \\\n",
    "--recode-INFO-all \\\n",
    "--remove {os.path.join(analysis_dir, \"bad_indv.txt\")} \\\n",
    "--out {os.path.join(analysis_dir, \"snps_indv_removed\")}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [],
   "source": [
    "vcf_bad_remove = os.path.join(analysis_dir, \"snps_indv_removed.recode.vcf\")\n",
    "vcf_bad_remove_gz = vcf_bad_remove + \".gz\"\n",
    "!$bgzip -c {vcf_bad_remove} > {vcf_bad_remove_gz}\n",
    "!$tabix {vcf_bad_remove_gz}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Filter snps further \n",
    "this needs to be done after removing individuals\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_vcf_stats(vcf_gz):\n",
    "    \n",
    "    stats = ['depth',\n",
    "            'site-depth',\n",
    "            'site-mean-depth',\n",
    "            'site-quality',\n",
    "            'missing-site',\n",
    "            'freq',\n",
    "            'counts',\n",
    "            'hardy',\n",
    "            'het']\n",
    "    \n",
    "    for stat in stats:\n",
    "        !$vcftools --gzvcf $vcf_gz \\\n",
    "        --out $vcf_gz \\\n",
    "        {\"--%s\" % stat} "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "VCFtools - 0.1.16\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf /data/gpfs/assoc/denovo/tfaske/sagebrush/SNPcall/filtering/snps_indv_removed.recode.vcf.gz\n",
      "\t--depth\n",
      "\t--out /data/gpfs/assoc/denovo/tfaske/sagebrush/SNPcall/filtering/snps_indv_removed.recode.vcf.gz\n",
      "\n",
      "Using zlib version: 1.2.11\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=PV4,Number=4,Type=Float,Description=\"P-values for strand bias, baseQ bias, mapQ bias and tail distance bias\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=PV4,Number=4,Type=Float,Description=\"P-values for strand bias, baseQ bias, mapQ bias and tail distance bias\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=DP4,Number=4,Type=Integer,Description=\"Number of high-quality ref-forward , ref-reverse, alt-forward and alt-reverse bases\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=DP4,Number=4,Type=Integer,Description=\"Number of high-quality ref-forward , ref-reverse, alt-forward and alt-reverse bases\">\n",
      "After filtering, kept 698 out of 698 Individuals\n",
      "Outputting Mean Depth by Individual\n",
      "After filtering, kept 46854 out of a possible 46854 Sites\n",
      "Run Time = 6.00 seconds\n",
      "\n",
      "VCFtools - 0.1.16\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf /data/gpfs/assoc/denovo/tfaske/sagebrush/SNPcall/filtering/snps_indv_removed.recode.vcf.gz\n",
      "\t--out /data/gpfs/assoc/denovo/tfaske/sagebrush/SNPcall/filtering/snps_indv_removed.recode.vcf.gz\n",
      "\t--site-depth\n",
      "\n",
      "Using zlib version: 1.2.11\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=PV4,Number=4,Type=Float,Description=\"P-values for strand bias, baseQ bias, mapQ bias and tail distance bias\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=PV4,Number=4,Type=Float,Description=\"P-values for strand bias, baseQ bias, mapQ bias and tail distance bias\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=DP4,Number=4,Type=Integer,Description=\"Number of high-quality ref-forward , ref-reverse, alt-forward and alt-reverse bases\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=DP4,Number=4,Type=Integer,Description=\"Number of high-quality ref-forward , ref-reverse, alt-forward and alt-reverse bases\">\n",
      "After filtering, kept 698 out of 698 Individuals\n",
      "Outputting Depth for Each Site\n",
      "After filtering, kept 46854 out of a possible 46854 Sites\n",
      "Run Time = 6.00 seconds\n",
      "\n",
      "VCFtools - 0.1.16\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf /data/gpfs/assoc/denovo/tfaske/sagebrush/SNPcall/filtering/snps_indv_removed.recode.vcf.gz\n",
      "\t--out /data/gpfs/assoc/denovo/tfaske/sagebrush/SNPcall/filtering/snps_indv_removed.recode.vcf.gz\n",
      "\t--site-mean-depth\n",
      "\n",
      "Using zlib version: 1.2.11\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=PV4,Number=4,Type=Float,Description=\"P-values for strand bias, baseQ bias, mapQ bias and tail distance bias\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=PV4,Number=4,Type=Float,Description=\"P-values for strand bias, baseQ bias, mapQ bias and tail distance bias\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=DP4,Number=4,Type=Integer,Description=\"Number of high-quality ref-forward , ref-reverse, alt-forward and alt-reverse bases\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=DP4,Number=4,Type=Integer,Description=\"Number of high-quality ref-forward , ref-reverse, alt-forward and alt-reverse bases\">\n",
      "After filtering, kept 698 out of 698 Individuals\n",
      "Outputting Depth for Each Site\n",
      "After filtering, kept 46854 out of a possible 46854 Sites\n",
      "Run Time = 6.00 seconds\n",
      "\n",
      "VCFtools - 0.1.16\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf /data/gpfs/assoc/denovo/tfaske/sagebrush/SNPcall/filtering/snps_indv_removed.recode.vcf.gz\n",
      "\t--out /data/gpfs/assoc/denovo/tfaske/sagebrush/SNPcall/filtering/snps_indv_removed.recode.vcf.gz\n",
      "\t--site-quality\n",
      "\n",
      "Using zlib version: 1.2.11\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=PV4,Number=4,Type=Float,Description=\"P-values for strand bias, baseQ bias, mapQ bias and tail distance bias\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=PV4,Number=4,Type=Float,Description=\"P-values for strand bias, baseQ bias, mapQ bias and tail distance bias\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=DP4,Number=4,Type=Integer,Description=\"Number of high-quality ref-forward , ref-reverse, alt-forward and alt-reverse bases\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=DP4,Number=4,Type=Integer,Description=\"Number of high-quality ref-forward , ref-reverse, alt-forward and alt-reverse bases\">\n",
      "After filtering, kept 698 out of 698 Individuals\n",
      "Outputting Quality for Each Site\n",
      "After filtering, kept 46854 out of a possible 46854 Sites\n",
      "Run Time = 2.00 seconds\n",
      "\n",
      "VCFtools - 0.1.16\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf /data/gpfs/assoc/denovo/tfaske/sagebrush/SNPcall/filtering/snps_indv_removed.recode.vcf.gz\n",
      "\t--out /data/gpfs/assoc/denovo/tfaske/sagebrush/SNPcall/filtering/snps_indv_removed.recode.vcf.gz\n",
      "\t--missing-site\n",
      "\n",
      "Using zlib version: 1.2.11\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=PV4,Number=4,Type=Float,Description=\"P-values for strand bias, baseQ bias, mapQ bias and tail distance bias\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=PV4,Number=4,Type=Float,Description=\"P-values for strand bias, baseQ bias, mapQ bias and tail distance bias\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=DP4,Number=4,Type=Integer,Description=\"Number of high-quality ref-forward , ref-reverse, alt-forward and alt-reverse bases\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=DP4,Number=4,Type=Integer,Description=\"Number of high-quality ref-forward , ref-reverse, alt-forward and alt-reverse bases\">\n",
      "After filtering, kept 698 out of 698 Individuals\n",
      "Outputting Site Missingness\n",
      "After filtering, kept 46854 out of a possible 46854 Sites\n",
      "Run Time = 6.00 seconds\n",
      "\n",
      "VCFtools - 0.1.16\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf /data/gpfs/assoc/denovo/tfaske/sagebrush/SNPcall/filtering/snps_indv_removed.recode.vcf.gz\n",
      "\t--freq\n",
      "\t--out /data/gpfs/assoc/denovo/tfaske/sagebrush/SNPcall/filtering/snps_indv_removed.recode.vcf.gz\n",
      "\n",
      "Using zlib version: 1.2.11\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=PV4,Number=4,Type=Float,Description=\"P-values for strand bias, baseQ bias, mapQ bias and tail distance bias\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=PV4,Number=4,Type=Float,Description=\"P-values for strand bias, baseQ bias, mapQ bias and tail distance bias\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=DP4,Number=4,Type=Integer,Description=\"Number of high-quality ref-forward , ref-reverse, alt-forward and alt-reverse bases\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=DP4,Number=4,Type=Integer,Description=\"Number of high-quality ref-forward , ref-reverse, alt-forward and alt-reverse bases\">\n",
      "After filtering, kept 698 out of 698 Individuals\n",
      "Outputting Frequency Statistics...\n",
      "After filtering, kept 46854 out of a possible 46854 Sites\n",
      "Run Time = 6.00 seconds\n",
      "\n",
      "VCFtools - 0.1.16\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf /data/gpfs/assoc/denovo/tfaske/sagebrush/SNPcall/filtering/snps_indv_removed.recode.vcf.gz\n",
      "\t--counts\n",
      "\t--out /data/gpfs/assoc/denovo/tfaske/sagebrush/SNPcall/filtering/snps_indv_removed.recode.vcf.gz\n",
      "\n",
      "Using zlib version: 1.2.11\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=PV4,Number=4,Type=Float,Description=\"P-values for strand bias, baseQ bias, mapQ bias and tail distance bias\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=PV4,Number=4,Type=Float,Description=\"P-values for strand bias, baseQ bias, mapQ bias and tail distance bias\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=DP4,Number=4,Type=Integer,Description=\"Number of high-quality ref-forward , ref-reverse, alt-forward and alt-reverse bases\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=DP4,Number=4,Type=Integer,Description=\"Number of high-quality ref-forward , ref-reverse, alt-forward and alt-reverse bases\">\n",
      "After filtering, kept 698 out of 698 Individuals\n",
      "Outputting Frequency Statistics...\n",
      "After filtering, kept 46854 out of a possible 46854 Sites\n",
      "Run Time = 5.00 seconds\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "VCFtools - 0.1.16\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf /data/gpfs/assoc/denovo/tfaske/sagebrush/SNPcall/filtering/snps_indv_removed.recode.vcf.gz\n",
      "\t--hardy\n",
      "\t--out /data/gpfs/assoc/denovo/tfaske/sagebrush/SNPcall/filtering/snps_indv_removed.recode.vcf.gz\n",
      "\n",
      "Using zlib version: 1.2.11\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=PV4,Number=4,Type=Float,Description=\"P-values for strand bias, baseQ bias, mapQ bias and tail distance bias\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=PV4,Number=4,Type=Float,Description=\"P-values for strand bias, baseQ bias, mapQ bias and tail distance bias\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=DP4,Number=4,Type=Integer,Description=\"Number of high-quality ref-forward , ref-reverse, alt-forward and alt-reverse bases\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=DP4,Number=4,Type=Integer,Description=\"Number of high-quality ref-forward , ref-reverse, alt-forward and alt-reverse bases\">\n",
      "After filtering, kept 698 out of 698 Individuals\n",
      "Outputting HWE statistics (but only for biallelic loci)\n",
      "After filtering, kept 46854 out of a possible 46854 Sites\n",
      "Run Time = 7.00 seconds\n",
      "\n",
      "VCFtools - 0.1.16\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf /data/gpfs/assoc/denovo/tfaske/sagebrush/SNPcall/filtering/snps_indv_removed.recode.vcf.gz\n",
      "\t--het\n",
      "\t--out /data/gpfs/assoc/denovo/tfaske/sagebrush/SNPcall/filtering/snps_indv_removed.recode.vcf.gz\n",
      "\n",
      "Using zlib version: 1.2.11\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=PV4,Number=4,Type=Float,Description=\"P-values for strand bias, baseQ bias, mapQ bias and tail distance bias\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=PV4,Number=4,Type=Float,Description=\"P-values for strand bias, baseQ bias, mapQ bias and tail distance bias\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=DP4,Number=4,Type=Integer,Description=\"Number of high-quality ref-forward , ref-reverse, alt-forward and alt-reverse bases\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=DP4,Number=4,Type=Integer,Description=\"Number of high-quality ref-forward , ref-reverse, alt-forward and alt-reverse bases\">\n",
      "After filtering, kept 698 out of 698 Individuals\n",
      "Outputting Individual Heterozygosity\n",
      "After filtering, kept 46854 out of a possible 46854 Sites\n",
      "Run Time = 5.00 seconds\n"
     ]
    }
   ],
   "source": [
    "get_vcf_stats(vcf_bad_remove_gz)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [],
   "source": [
    "#pd.set_option('display.max_columns', 100)\n",
    "\n",
    "def get_MAF(row):\n",
    "    try:\n",
    "        return np.min([row.A1_freq, row.A2_freq])\n",
    "    except:\n",
    "        print(row)\n",
    "        \n",
    "def get_correction(n):\n",
    "    #for finite sample size\n",
    "    return (2*n)/(2*n-1)\n",
    "\n",
    "def calculate_Fis(vals):\n",
    "    try:\n",
    "        data = [float(x) for x in vals.split(\"/\")]\n",
    "        assert len(data) == 3\n",
    "        num_individuals = np.sum(data)\n",
    "        total_alleles = 2*num_individuals\n",
    "        a1_count = 2*data[0]\n",
    "        a2_count = 2*data[2]\n",
    "        het_count = data[1]\n",
    "        a1_count += het_count\n",
    "        a2_count += het_count\n",
    "        a1_freq = a1_count/total_alleles\n",
    "        a2_freq = a2_count/total_alleles\n",
    "        assert a1_freq + a2_freq == 1.0\n",
    "        He = 2 * a1_freq * a2_freq * get_correction(num_individuals)\n",
    "        Ho = het_count/num_individuals\n",
    "        Fis = 1 - (Ho/He)\n",
    "        return Fis\n",
    "    except:\n",
    "        return -9\n",
    "\n",
    "def combine_vcf_stats(filedir, prefix):\n",
    "    \n",
    "    hardy_files = !ls {filedir}/{prefix}.hwe\n",
    "    hardy = pd.read_csv(hardy_files[0], sep=\"\\t\")\n",
    "\n",
    "    hardy.columns = ['CHROM', 'POS', 'OBS(HOM1/HET/HOM2)', 'E(HOM1/HET/HOM2)', 'ChiSq_HWE',\n",
    "       'P_HWE', 'P_HET_DEFICIT', 'P_HET_EXCESS']\n",
    "    hardy.index = hardy.apply(lambda x: \"%s-%d\" % (x.CHROM, x.POS), axis=1)\n",
    "    \n",
    "    loci_files = !ls {filedir}/{prefix}.l* | grep -v log\n",
    "    loci_df = pd.concat([pd.read_csv(x, sep=\"\\t\", skiprows=0) for x in loci_files], axis=1)\n",
    "    chrom_pos = loci_df.ix[:,0:2]\n",
    "    \n",
    "    frq_files = !ls {filedir}/{prefix}.frq* | grep -v count\n",
    "    frq_data = []\n",
    "    h = open(frq_files[0])\n",
    "    header = h.readline().strip().split()\n",
    "    for line in h:\n",
    "        frq_data.append(line.strip().split('\\t'))\n",
    "\n",
    "    header = ['CHROM', 'POS', 'N_ALLELES', 'N_CHR', 'A1_FREQ', \"A2_FREQ\"]\n",
    "    frq_df = pd.DataFrame(frq_data)\n",
    "    print(frq_df.columns)\n",
    "    #frq_df = frq_df.drop([6,7],axis=1)\n",
    "    frq_df.columns = header\n",
    "    frq_df.index = frq_df.apply(lambda x: \"%s-%s\" % (x.CHROM, x.POS), axis=1)\n",
    "    \n",
    "    loci_df = loci_df.drop(['CHROM','CHR','POS'], axis=1)\n",
    "    loci_df = pd.concat([chrom_pos, loci_df], axis=1)\n",
    "    loci_df.index = loci_df.apply(lambda x: \"%s-%d\" % (x.CHROM, x.POS), axis=1)\n",
    "    \n",
    "    loci_df = pd.concat([loci_df, frq_df, hardy], axis=1)\n",
    "    loci_df[\"A1_allele\"] = loci_df.apply(lambda row: row.A1_FREQ.split(\":\")[0], axis=1)\n",
    "    loci_df[\"A2_allele\"] = loci_df.apply(lambda row: row.A2_FREQ.split(\":\")[0], axis=1)\n",
    "    \n",
    "    loci_df[\"A1_freq\"] = loci_df.apply(lambda row: float(row.A1_FREQ.split(\":\")[1]), axis=1)\n",
    "    loci_df[\"A2_freq\"] = loci_df.apply(lambda row: float(row.A2_FREQ.split(\":\")[1]), axis=1)\n",
    "    \n",
    "    loci_df['MAF'] = loci_df.apply(get_MAF, axis=1)\n",
    "    loci_df = loci_df.drop(['CHROM', 'POS'], axis=1)\n",
    "    \n",
    "    loci_df['Fis'] = loci_df['OBS(HOM1/HET/HOM2)'].apply(calculate_Fis)\n",
    "    \n",
    "    return loci_df, frq_df, hardy"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'/data/gpfs/assoc/denovo/tfaske/sagebrush/SNPcall/filtering/snps_indv_removed.recode.vcf.gz'"
      ]
     },
     "execution_count": 52,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "vcf_bad_remove_gz"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/data/gpfs/assoc/parchmanlab/tfaske/anaconda3/envs/py36/lib/python3.6/site-packages/ipykernel_launcher.py:45: FutureWarning: \n",
      ".ix is deprecated. Please use\n",
      ".loc for label based indexing or\n",
      ".iloc for positional indexing\n",
      "\n",
      "See the documentation here:\n",
      "http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#ix-indexer-is-deprecated\n",
      "/data/gpfs/assoc/parchmanlab/tfaske/anaconda3/envs/py36/lib/python3.6/site-packages/pandas/core/indexing.py:822: FutureWarning: \n",
      ".ix is deprecated. Please use\n",
      ".loc for label based indexing or\n",
      ".iloc for positional indexing\n",
      "\n",
      "See the documentation here:\n",
      "http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#ix-indexer-is-deprecated\n",
      "  retval = getattr(retval, self.name)._getitem_axis(key, axis=i)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "RangeIndex(start=0, stop=6, step=1)\n"
     ]
    }
   ],
   "source": [
    "loci_df, frq_df, hardy = combine_vcf_stats(analysis_dir,'snps_indv_removed.recode.vcf.gz')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SUM_DEPTH</th>\n",
       "      <th>SUMSQ_DEPTH</th>\n",
       "      <th>MEAN_DEPTH</th>\n",
       "      <th>VAR_DEPTH</th>\n",
       "      <th>N_DATA</th>\n",
       "      <th>N_GENOTYPE_FILTERED</th>\n",
       "      <th>N_MISS</th>\n",
       "      <th>F_MISS</th>\n",
       "      <th>QUAL</th>\n",
       "      <th>N_ALLELES</th>\n",
       "      <th>...</th>\n",
       "      <th>ChiSq_HWE</th>\n",
       "      <th>P_HWE</th>\n",
       "      <th>P_HET_DEFICIT</th>\n",
       "      <th>P_HET_EXCESS</th>\n",
       "      <th>A1_allele</th>\n",
       "      <th>A2_allele</th>\n",
       "      <th>A1_freq</th>\n",
       "      <th>A2_freq</th>\n",
       "      <th>MAF</th>\n",
       "      <th>Fis</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>CM042338.1-126072</th>\n",
       "      <td>555</td>\n",
       "      <td>1601</td>\n",
       "      <td>0.795129</td>\n",
       "      <td>1.66385</td>\n",
       "      <td>1396</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>999.0</td>\n",
       "      <td>2</td>\n",
       "      <td>...</td>\n",
       "      <td>61.682280</td>\n",
       "      <td>2.580264e-09</td>\n",
       "      <td>2.580264e-09</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>0.933381</td>\n",
       "      <td>0.066619</td>\n",
       "      <td>0.066619</td>\n",
       "      <td>0.297774</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CM042338.1-1342981</th>\n",
       "      <td>1128</td>\n",
       "      <td>4994</td>\n",
       "      <td>1.616050</td>\n",
       "      <td>4.54964</td>\n",
       "      <td>1396</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>999.0</td>\n",
       "      <td>2</td>\n",
       "      <td>...</td>\n",
       "      <td>9.918258</td>\n",
       "      <td>2.855945e-03</td>\n",
       "      <td>2.292476e-03</td>\n",
       "      <td>0.999105</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.866762</td>\n",
       "      <td>0.133238</td>\n",
       "      <td>0.133238</td>\n",
       "      <td>0.119835</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CM042338.1-1496841</th>\n",
       "      <td>2066</td>\n",
       "      <td>19146</td>\n",
       "      <td>2.959890</td>\n",
       "      <td>18.69570</td>\n",
       "      <td>1396</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>999.0</td>\n",
       "      <td>2</td>\n",
       "      <td>...</td>\n",
       "      <td>0.082348</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>0.926811</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>0.989255</td>\n",
       "      <td>0.010745</td>\n",
       "      <td>0.010745</td>\n",
       "      <td>-0.010138</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CM042338.1-1744262</th>\n",
       "      <td>1081</td>\n",
       "      <td>4957</td>\n",
       "      <td>1.548710</td>\n",
       "      <td>4.70996</td>\n",
       "      <td>1396</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>999.0</td>\n",
       "      <td>2</td>\n",
       "      <td>...</td>\n",
       "      <td>0.119097</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>0.894945</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0.987106</td>\n",
       "      <td>0.012894</td>\n",
       "      <td>0.012894</td>\n",
       "      <td>-0.012337</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CM042338.1-1942811</th>\n",
       "      <td>1673</td>\n",
       "      <td>14005</td>\n",
       "      <td>2.396850</td>\n",
       "      <td>14.34010</td>\n",
       "      <td>1396</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>999.0</td>\n",
       "      <td>2</td>\n",
       "      <td>...</td>\n",
       "      <td>0.697515</td>\n",
       "      <td>6.113742e-01</td>\n",
       "      <td>8.618386e-01</td>\n",
       "      <td>0.300844</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.918338</td>\n",
       "      <td>0.081662</td>\n",
       "      <td>0.081662</td>\n",
       "      <td>-0.030873</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CM042346.1-338104939</th>\n",
       "      <td>1753</td>\n",
       "      <td>8185</td>\n",
       "      <td>2.511460</td>\n",
       "      <td>5.42670</td>\n",
       "      <td>1396</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>999.0</td>\n",
       "      <td>2</td>\n",
       "      <td>...</td>\n",
       "      <td>0.132891</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>0.883255</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>0.986390</td>\n",
       "      <td>0.013610</td>\n",
       "      <td>0.013610</td>\n",
       "      <td>-0.013072</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CM042346.1-338112942</th>\n",
       "      <td>2307</td>\n",
       "      <td>21957</td>\n",
       "      <td>3.305160</td>\n",
       "      <td>20.56240</td>\n",
       "      <td>1396</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>999.0</td>\n",
       "      <td>2</td>\n",
       "      <td>...</td>\n",
       "      <td>0.106078</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>0.906115</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>0.987822</td>\n",
       "      <td>0.012178</td>\n",
       "      <td>0.012178</td>\n",
       "      <td>-0.011603</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CM042346.1-338202655</th>\n",
       "      <td>1578</td>\n",
       "      <td>6784</td>\n",
       "      <td>2.260740</td>\n",
       "      <td>4.61484</td>\n",
       "      <td>1396</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>999.0</td>\n",
       "      <td>2</td>\n",
       "      <td>...</td>\n",
       "      <td>0.119097</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>0.894945</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>0.987106</td>\n",
       "      <td>0.012894</td>\n",
       "      <td>0.012894</td>\n",
       "      <td>-0.012337</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CM042346.1-338204118</th>\n",
       "      <td>1710</td>\n",
       "      <td>11438</td>\n",
       "      <td>2.449860</td>\n",
       "      <td>10.39990</td>\n",
       "      <td>1396</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>999.0</td>\n",
       "      <td>2</td>\n",
       "      <td>...</td>\n",
       "      <td>1.003578</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>0.387583</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0.963467</td>\n",
       "      <td>0.036533</td>\n",
       "      <td>0.036533</td>\n",
       "      <td>-0.037175</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CM042346.1-338258033</th>\n",
       "      <td>1622</td>\n",
       "      <td>8874</td>\n",
       "      <td>2.323780</td>\n",
       "      <td>7.32400</td>\n",
       "      <td>1396</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>999.0</td>\n",
       "      <td>2</td>\n",
       "      <td>...</td>\n",
       "      <td>0.671609</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>0.529534</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0.969914</td>\n",
       "      <td>0.030086</td>\n",
       "      <td>0.030086</td>\n",
       "      <td>-0.030281</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>46854 rows  25 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                      SUM_DEPTH  SUMSQ_DEPTH  MEAN_DEPTH  VAR_DEPTH  N_DATA  \\\n",
       "CM042338.1-126072           555         1601    0.795129    1.66385    1396   \n",
       "CM042338.1-1342981         1128         4994    1.616050    4.54964    1396   \n",
       "CM042338.1-1496841         2066        19146    2.959890   18.69570    1396   \n",
       "CM042338.1-1744262         1081         4957    1.548710    4.70996    1396   \n",
       "CM042338.1-1942811         1673        14005    2.396850   14.34010    1396   \n",
       "...                         ...          ...         ...        ...     ...   \n",
       "CM042346.1-338104939       1753         8185    2.511460    5.42670    1396   \n",
       "CM042346.1-338112942       2307        21957    3.305160   20.56240    1396   \n",
       "CM042346.1-338202655       1578         6784    2.260740    4.61484    1396   \n",
       "CM042346.1-338204118       1710        11438    2.449860   10.39990    1396   \n",
       "CM042346.1-338258033       1622         8874    2.323780    7.32400    1396   \n",
       "\n",
       "                      N_GENOTYPE_FILTERED  N_MISS  F_MISS   QUAL N_ALLELES  \\\n",
       "CM042338.1-126072                       0       0       0  999.0         2   \n",
       "CM042338.1-1342981                      0       0       0  999.0         2   \n",
       "CM042338.1-1496841                      0       0       0  999.0         2   \n",
       "CM042338.1-1744262                      0       0       0  999.0         2   \n",
       "CM042338.1-1942811                      0       0       0  999.0         2   \n",
       "...                                   ...     ...     ...    ...       ...   \n",
       "CM042346.1-338104939                    0       0       0  999.0         2   \n",
       "CM042346.1-338112942                    0       0       0  999.0         2   \n",
       "CM042346.1-338202655                    0       0       0  999.0         2   \n",
       "CM042346.1-338204118                    0       0       0  999.0         2   \n",
       "CM042346.1-338258033                    0       0       0  999.0         2   \n",
       "\n",
       "                      ...  ChiSq_HWE         P_HWE P_HET_DEFICIT P_HET_EXCESS  \\\n",
       "CM042338.1-126072     ...  61.682280  2.580264e-09  2.580264e-09     1.000000   \n",
       "CM042338.1-1342981    ...   9.918258  2.855945e-03  2.292476e-03     0.999105   \n",
       "CM042338.1-1496841    ...   0.082348  1.000000e+00  1.000000e+00     0.926811   \n",
       "CM042338.1-1744262    ...   0.119097  1.000000e+00  1.000000e+00     0.894945   \n",
       "CM042338.1-1942811    ...   0.697515  6.113742e-01  8.618386e-01     0.300844   \n",
       "...                   ...        ...           ...           ...          ...   \n",
       "CM042346.1-338104939  ...   0.132891  1.000000e+00  1.000000e+00     0.883255   \n",
       "CM042346.1-338112942  ...   0.106078  1.000000e+00  1.000000e+00     0.906115   \n",
       "CM042346.1-338202655  ...   0.119097  1.000000e+00  1.000000e+00     0.894945   \n",
       "CM042346.1-338204118  ...   1.003578  1.000000e+00  1.000000e+00     0.387583   \n",
       "CM042346.1-338258033  ...   0.671609  1.000000e+00  1.000000e+00     0.529534   \n",
       "\n",
       "                     A1_allele  A2_allele   A1_freq   A2_freq       MAF  \\\n",
       "CM042338.1-126072            C          G  0.933381  0.066619  0.066619   \n",
       "CM042338.1-1342981           G          A  0.866762  0.133238  0.133238   \n",
       "CM042338.1-1496841           G          C  0.989255  0.010745  0.010745   \n",
       "CM042338.1-1744262           C          T  0.987106  0.012894  0.012894   \n",
       "CM042338.1-1942811           G          A  0.918338  0.081662  0.081662   \n",
       "...                        ...        ...       ...       ...       ...   \n",
       "CM042346.1-338104939         C          A  0.986390  0.013610  0.013610   \n",
       "CM042346.1-338112942         T          C  0.987822  0.012178  0.012178   \n",
       "CM042346.1-338202655         T          G  0.987106  0.012894  0.012894   \n",
       "CM042346.1-338204118         C          T  0.963467  0.036533  0.036533   \n",
       "CM042346.1-338258033         C          T  0.969914  0.030086  0.030086   \n",
       "\n",
       "                           Fis  \n",
       "CM042338.1-126072     0.297774  \n",
       "CM042338.1-1342981    0.119835  \n",
       "CM042338.1-1496841   -0.010138  \n",
       "CM042338.1-1744262   -0.012337  \n",
       "CM042338.1-1942811   -0.030873  \n",
       "...                        ...  \n",
       "CM042346.1-338104939 -0.013072  \n",
       "CM042346.1-338112942 -0.011603  \n",
       "CM042346.1-338202655 -0.012337  \n",
       "CM042346.1-338204118 -0.037175  \n",
       "CM042346.1-338258033 -0.030281  \n",
       "\n",
       "[46854 rows x 25 columns]"
      ]
     },
     "execution_count": 54,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "loci_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHROM</th>\n",
       "      <th>POS</th>\n",
       "      <th>N_ALLELES</th>\n",
       "      <th>N_CHR</th>\n",
       "      <th>A1_FREQ</th>\n",
       "      <th>A2_FREQ</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>CM042338.1-126072</th>\n",
       "      <td>CM042338.1</td>\n",
       "      <td>126072</td>\n",
       "      <td>2</td>\n",
       "      <td>1396</td>\n",
       "      <td>C:0.933381</td>\n",
       "      <td>G:0.0666189</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CM042338.1-1342981</th>\n",
       "      <td>CM042338.1</td>\n",
       "      <td>1342981</td>\n",
       "      <td>2</td>\n",
       "      <td>1396</td>\n",
       "      <td>G:0.866762</td>\n",
       "      <td>A:0.133238</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CM042338.1-1496841</th>\n",
       "      <td>CM042338.1</td>\n",
       "      <td>1496841</td>\n",
       "      <td>2</td>\n",
       "      <td>1396</td>\n",
       "      <td>G:0.989255</td>\n",
       "      <td>C:0.010745</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CM042338.1-1744262</th>\n",
       "      <td>CM042338.1</td>\n",
       "      <td>1744262</td>\n",
       "      <td>2</td>\n",
       "      <td>1396</td>\n",
       "      <td>C:0.987106</td>\n",
       "      <td>T:0.012894</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CM042338.1-1942811</th>\n",
       "      <td>CM042338.1</td>\n",
       "      <td>1942811</td>\n",
       "      <td>2</td>\n",
       "      <td>1396</td>\n",
       "      <td>G:0.918338</td>\n",
       "      <td>A:0.0816619</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CM042346.1-338104939</th>\n",
       "      <td>CM042346.1</td>\n",
       "      <td>338104939</td>\n",
       "      <td>2</td>\n",
       "      <td>1396</td>\n",
       "      <td>C:0.98639</td>\n",
       "      <td>A:0.0136103</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CM042346.1-338112942</th>\n",
       "      <td>CM042346.1</td>\n",
       "      <td>338112942</td>\n",
       "      <td>2</td>\n",
       "      <td>1396</td>\n",
       "      <td>T:0.987822</td>\n",
       "      <td>C:0.0121777</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CM042346.1-338202655</th>\n",
       "      <td>CM042346.1</td>\n",
       "      <td>338202655</td>\n",
       "      <td>2</td>\n",
       "      <td>1396</td>\n",
       "      <td>T:0.987106</td>\n",
       "      <td>G:0.012894</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CM042346.1-338204118</th>\n",
       "      <td>CM042346.1</td>\n",
       "      <td>338204118</td>\n",
       "      <td>2</td>\n",
       "      <td>1396</td>\n",
       "      <td>C:0.963467</td>\n",
       "      <td>T:0.036533</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CM042346.1-338258033</th>\n",
       "      <td>CM042346.1</td>\n",
       "      <td>338258033</td>\n",
       "      <td>2</td>\n",
       "      <td>1396</td>\n",
       "      <td>C:0.969914</td>\n",
       "      <td>T:0.030086</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>46854 rows  6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                           CHROM        POS N_ALLELES N_CHR     A1_FREQ  \\\n",
       "CM042338.1-126072     CM042338.1     126072         2  1396  C:0.933381   \n",
       "CM042338.1-1342981    CM042338.1    1342981         2  1396  G:0.866762   \n",
       "CM042338.1-1496841    CM042338.1    1496841         2  1396  G:0.989255   \n",
       "CM042338.1-1744262    CM042338.1    1744262         2  1396  C:0.987106   \n",
       "CM042338.1-1942811    CM042338.1    1942811         2  1396  G:0.918338   \n",
       "...                          ...        ...       ...   ...         ...   \n",
       "CM042346.1-338104939  CM042346.1  338104939         2  1396   C:0.98639   \n",
       "CM042346.1-338112942  CM042346.1  338112942         2  1396  T:0.987822   \n",
       "CM042346.1-338202655  CM042346.1  338202655         2  1396  T:0.987106   \n",
       "CM042346.1-338204118  CM042346.1  338204118         2  1396  C:0.963467   \n",
       "CM042346.1-338258033  CM042346.1  338258033         2  1396  C:0.969914   \n",
       "\n",
       "                          A2_FREQ  \n",
       "CM042338.1-126072     G:0.0666189  \n",
       "CM042338.1-1342981     A:0.133238  \n",
       "CM042338.1-1496841     C:0.010745  \n",
       "CM042338.1-1744262     T:0.012894  \n",
       "CM042338.1-1942811    A:0.0816619  \n",
       "...                           ...  \n",
       "CM042346.1-338104939  A:0.0136103  \n",
       "CM042346.1-338112942  C:0.0121777  \n",
       "CM042346.1-338202655   G:0.012894  \n",
       "CM042346.1-338204118   T:0.036533  \n",
       "CM042346.1-338258033   T:0.030086  \n",
       "\n",
       "[46854 rows x 6 columns]"
      ]
     },
     "execution_count": 55,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "frq_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHROM</th>\n",
       "      <th>POS</th>\n",
       "      <th>OBS(HOM1/HET/HOM2)</th>\n",
       "      <th>E(HOM1/HET/HOM2)</th>\n",
       "      <th>ChiSq_HWE</th>\n",
       "      <th>P_HWE</th>\n",
       "      <th>P_HET_DEFICIT</th>\n",
       "      <th>P_HET_EXCESS</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>CM042338.1-126072</th>\n",
       "      <td>CM042338.1</td>\n",
       "      <td>126072</td>\n",
       "      <td>621/61/16</td>\n",
       "      <td>608.10/86.80/3.10</td>\n",
       "      <td>61.682280</td>\n",
       "      <td>2.580264e-09</td>\n",
       "      <td>2.580264e-09</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CM042338.1-1342981</th>\n",
       "      <td>CM042338.1</td>\n",
       "      <td>1342981</td>\n",
       "      <td>534/142/22</td>\n",
       "      <td>524.39/161.22/12.39</td>\n",
       "      <td>9.918258</td>\n",
       "      <td>2.855945e-03</td>\n",
       "      <td>2.292476e-03</td>\n",
       "      <td>0.999105</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CM042338.1-1496841</th>\n",
       "      <td>CM042338.1</td>\n",
       "      <td>1496841</td>\n",
       "      <td>683/15/0</td>\n",
       "      <td>683.08/14.84/0.08</td>\n",
       "      <td>0.082348</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>0.926811</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CM042338.1-1744262</th>\n",
       "      <td>CM042338.1</td>\n",
       "      <td>1744262</td>\n",
       "      <td>680/18/0</td>\n",
       "      <td>680.12/17.77/0.12</td>\n",
       "      <td>0.119097</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>0.894945</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CM042338.1-1942811</th>\n",
       "      <td>CM042338.1</td>\n",
       "      <td>1942811</td>\n",
       "      <td>587/108/3</td>\n",
       "      <td>588.65/104.69/4.65</td>\n",
       "      <td>0.697515</td>\n",
       "      <td>6.113742e-01</td>\n",
       "      <td>8.618386e-01</td>\n",
       "      <td>0.300844</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CM042346.1-338104939</th>\n",
       "      <td>CM042346.1</td>\n",
       "      <td>338104939</td>\n",
       "      <td>679/19/0</td>\n",
       "      <td>679.13/18.74/0.13</td>\n",
       "      <td>0.132891</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>0.883255</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CM042346.1-338112942</th>\n",
       "      <td>CM042346.1</td>\n",
       "      <td>338112942</td>\n",
       "      <td>681/17/0</td>\n",
       "      <td>681.10/16.79/0.10</td>\n",
       "      <td>0.106078</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>0.906115</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CM042346.1-338202655</th>\n",
       "      <td>CM042346.1</td>\n",
       "      <td>338202655</td>\n",
       "      <td>680/18/0</td>\n",
       "      <td>680.12/17.77/0.12</td>\n",
       "      <td>0.119097</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>0.894945</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CM042346.1-338204118</th>\n",
       "      <td>CM042346.1</td>\n",
       "      <td>338204118</td>\n",
       "      <td>647/51/0</td>\n",
       "      <td>647.93/49.14/0.93</td>\n",
       "      <td>1.003578</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>0.387583</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CM042346.1-338258033</th>\n",
       "      <td>CM042346.1</td>\n",
       "      <td>338258033</td>\n",
       "      <td>656/42/0</td>\n",
       "      <td>656.63/40.74/0.63</td>\n",
       "      <td>0.671609</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>0.529534</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>46854 rows  8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                           CHROM        POS OBS(HOM1/HET/HOM2)  \\\n",
       "CM042338.1-126072     CM042338.1     126072          621/61/16   \n",
       "CM042338.1-1342981    CM042338.1    1342981         534/142/22   \n",
       "CM042338.1-1496841    CM042338.1    1496841           683/15/0   \n",
       "CM042338.1-1744262    CM042338.1    1744262           680/18/0   \n",
       "CM042338.1-1942811    CM042338.1    1942811          587/108/3   \n",
       "...                          ...        ...                ...   \n",
       "CM042346.1-338104939  CM042346.1  338104939           679/19/0   \n",
       "CM042346.1-338112942  CM042346.1  338112942           681/17/0   \n",
       "CM042346.1-338202655  CM042346.1  338202655           680/18/0   \n",
       "CM042346.1-338204118  CM042346.1  338204118           647/51/0   \n",
       "CM042346.1-338258033  CM042346.1  338258033           656/42/0   \n",
       "\n",
       "                         E(HOM1/HET/HOM2)  ChiSq_HWE         P_HWE  \\\n",
       "CM042338.1-126072       608.10/86.80/3.10  61.682280  2.580264e-09   \n",
       "CM042338.1-1342981    524.39/161.22/12.39   9.918258  2.855945e-03   \n",
       "CM042338.1-1496841      683.08/14.84/0.08   0.082348  1.000000e+00   \n",
       "CM042338.1-1744262      680.12/17.77/0.12   0.119097  1.000000e+00   \n",
       "CM042338.1-1942811     588.65/104.69/4.65   0.697515  6.113742e-01   \n",
       "...                                   ...        ...           ...   \n",
       "CM042346.1-338104939    679.13/18.74/0.13   0.132891  1.000000e+00   \n",
       "CM042346.1-338112942    681.10/16.79/0.10   0.106078  1.000000e+00   \n",
       "CM042346.1-338202655    680.12/17.77/0.12   0.119097  1.000000e+00   \n",
       "CM042346.1-338204118    647.93/49.14/0.93   1.003578  1.000000e+00   \n",
       "CM042346.1-338258033    656.63/40.74/0.63   0.671609  1.000000e+00   \n",
       "\n",
       "                      P_HET_DEFICIT  P_HET_EXCESS  \n",
       "CM042338.1-126072      2.580264e-09      1.000000  \n",
       "CM042338.1-1342981     2.292476e-03      0.999105  \n",
       "CM042338.1-1496841     1.000000e+00      0.926811  \n",
       "CM042338.1-1744262     1.000000e+00      0.894945  \n",
       "CM042338.1-1942811     8.618386e-01      0.300844  \n",
       "...                             ...           ...  \n",
       "CM042346.1-338104939   1.000000e+00      0.883255  \n",
       "CM042346.1-338112942   1.000000e+00      0.906115  \n",
       "CM042346.1-338202655   1.000000e+00      0.894945  \n",
       "CM042346.1-338204118   1.000000e+00      0.387583  \n",
       "CM042346.1-338258033   1.000000e+00      0.529534  \n",
       "\n",
       "[46854 rows x 8 columns]"
      ]
     },
     "execution_count": 56,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "hardy"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {},
   "outputs": [],
   "source": [
    "loci_out = os.path.join(analysis_dir,'loci_out.csv')\n",
    "frq_out = os.path.join(analysis_dir,'frq_out.csv')\n",
    "hardy_out = os.path.join(analysis_dir,'hardy_out.csv')\n",
    "\n",
    "loci_df.to_csv(path_or_buf=loci_out)\n",
    "frq_df.to_csv(path_or_buf=frq_out)\n",
    "hardy.to_csv(path_or_buf=hardy_out)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "count     46854.000000\n",
       "mean       1451.569407\n",
       "std        1801.784653\n",
       "min         418.000000\n",
       "25%         819.000000\n",
       "50%        1152.000000\n",
       "75%        1711.000000\n",
       "max      163984.000000\n",
       "Name: SUM_DEPTH, dtype: float64"
      ]
     },
     "execution_count": 58,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "loci_df.SUM_DEPTH.describe()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "count    46854.000000\n",
       "mean         2.079612\n",
       "std          2.581353\n",
       "min          0.598854\n",
       "25%          1.173350\n",
       "50%          1.650430\n",
       "75%          2.451290\n",
       "max        234.934000\n",
       "Name: MEAN_DEPTH, dtype: float64"
      ]
     },
     "execution_count": 59,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "loci_df.MEAN_DEPTH.describe()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "count    46854.000000\n",
       "mean       995.497164\n",
       "std         55.662527\n",
       "min         30.374800\n",
       "25%        999.000000\n",
       "50%        999.000000\n",
       "75%        999.000000\n",
       "max        999.000000\n",
       "Name: QUAL, dtype: float64"
      ]
     },
     "execution_count": 60,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "loci_df.QUAL.describe()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAyMAAAHwCAYAAAC15rBtAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAAWJQAAFiUBSVIk8AAAADh0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uMy4yLjEsIGh0dHA6Ly9tYXRwbG90bGliLm9yZy+j8jraAAAgAElEQVR4nO3df7xtVV03+s9XCFBUQNREsQuaKE/WVcF+HB8R9WZapKaU+Dwp/q4ewUT0MX+jYfkkiiImN0pPaV1MTH0w/FWIiMdUwC52JVHhZCrmD+AgcECPjPvHnLtWi73P2fuwzh6sfd7v12u+JmvOMeYYe032Ouuz5xxzVGstAAAAq+02vTsAAADsnIQRAACgC2EEAADoQhgBAAC6EEYAAIAuhBEAAKALYQQAAOhCGAEAALoQRgAAgC6EEQAAoAthBAAA6EIYAQAAuti1dwfYcarq8iR3TLKxc1cAAFi7DkhyTWvtwJVWFEbWtjve9ra3vdPBBx98p94dAQBgbbrkkkuyefPm7aorjKxtGw8++OA7XXjhhb37AQDAGnXIIYfkoosu2rg9dY0ZAQAAuhBGAACALoQRAACgC2EEAADoQhgBAAC6EEYAAIAuhBEAAKALYQQAAOhCGAEAALoQRgAAgC6EEQAAoAthBAAA6EIYAQAAuhBGAACALoQRAACgC2EEAADoQhgBAAC6EEYAAIAudu3dAQAA2BEO+L2/7d2FVbfxdb/Suwsr4soIAADQhTACAAB0IYwAAABdCCMAAEAXwggAANCFMAIAAHQhjAAAAF0IIwAAQBfCCAAA0IUwAgAAdCGMAAAAXQgjAABAF8IIAADQhTACAAB0IYwAAABdCCMAAEAXwggAANCFMAIAAHQhjAAAAF0IIwAAQBfCCAAA0IUwAgAAdCGMAAAAXQgjAABAF8IIAADQhTACAAB0IYwAAABdCCMAAEAXwggAANCFMAIAAHQhjAAAAF0IIwAAQBfCCAAA0IUwAgAAdCGMAAAAXQgjAABAF8IIAADQhTACAAB0IYwAAABdCCMAAEAXwggAANCFMAIAAHQhjAAAAF0IIwAAQBfCCAAA0IUwAgAAdCGMAAAAXQgjAABAF8IIAADQhTACAAB0IYwAAABdCCMAAEAXwggAANDF3IeRqvrNqmrj8qwlyhxRVedW1aaquraqPlNVR2/juEdX1WfH8pvG+kdspfwuVXVcVV1cVZur6sqqOruq1m2lzm2r6tVV9aWquqGqvl1Vf11VBy//HQAAgPk012Gkqu6Z5NQk126lzDFJzkpy/yTvSnJ6krsnWV9VJy1R56Qk65PsN5Z/V5KfTnLWeLzp8pXkjCRvTLLb2Kf3JTksyXlV9bhF6uye5GNJXpnkmiRvTvJ3SX4tyQVV9XPbfAMAAGCO7dq7A9trDADvSPK9JH+T5IWLlDkgyUlJrkxyaGtt47j9NUk+l+T4qnpva+3TE3XWJTk+yVeTPLi1dtW4/fVJLkxyUlV9cOFYo6OSHJlkQ5JHttZuGOucluT8JKdX1Tmtte9P1HlBkockOTPJk1prN4113p3k/UneXlU/vbAdAADWmnm+MvK8JI9I8vQk1y1R5hlJdk9y6mR4GAPGH4wvf3uqzsLr1y4EkbHOxiRvHY/39Kk6vzOuX74QRMY6n0vy7iR3yRBWkvx7kFpo539OBo7W2geSfDLJf0nysCV+LgAAmHtzGUbGMRWvS/Lm1tp5Wyn6iHH94UX2fWiqzHbVqao9kqxLcn2GELGcdu6d5CeSXNpau3wFfQMAgDVj7m7Tqqpdk7wzydeSvHQbxe87ri+d3tFau6Kqrkuyf1XdrrV2fVXtmeQeSa5trV2xyPG+PK4Pmth27yS7JLmstbZlmXWW7NdW6iypqi5cYtf9llMfAAB6mLswkmHA9wOT/NfW2uZtlN1rXG9aYv+mJHuO5a5fZvkk2XuFbcyiDgAArClzFUbGJ0y9NMkbJged7+xaa4cstn28YvKgVe4OAAAsy9yMGRlvz/qLDLc2vWKZ1RauMOy1xP7pKxTLLX/1drRxS+sAAMCaMjdhJMntM4yhODjJDRMTHbYkrxrLnD5ue9P4+kvj+mZjL6pqvwy3aH29tXZ9krTWrkvyjSS3H/dPu8+4nhzr8dUkP0pyrzEwLafOkv3aSh0AAFhT5uk2rRuT/NkS+x6UYRzJ+Rm+6C/cwnVOhrk8Hj2xbcFjJspMOifJU8Y679hWndbaDVW1IclDx+Xjy2jnqxkG4B9UVQcu8kStpfoGAABrxtxcGWmtbW6tPWuxJcn/Hov9+bjt3ePrd2QIMceMEyAmSapqn/zHk7hOm2pq4fXLxnILdQ5I8tzxeNMh5W3j+sTxUb8LdR6c5ElJvpPkvRM/S5to54+q6jYTdR6XIdR8Mcknln5HAABgvs3TlZEVa61dXlUvSnJKkgvG2c1/kGECwv2zyED41tqGqnpjhhnSL66qM5PsliFU3CnJsVOzryfJGUmeMB7381V1VpJ9xzq7JHl2a+2aqTpvTHLEWOczVfX3GeYe+fUMT/Z6htnXAQBYy9Z0GEmS1tpbqmpjkhcmeWqGq0FfzDBb+p8vUef4qvpChishz0lyU5KLkry+tfbBRcq3qnpykg0ZZn0/NskNSc5LcmJrbcMidW6sql9M8ntJnpzkuCTXJHl/kle11r54i35wAAC4lVsTYaS1dkKSE7ay/6wkZ63wmOuTrF9B+S1JTh6X5da5PsO8Ka9cSd8AAGAtmJsxIwAAwNoijAAAAF0IIwAAQBfCCAAA0IUwAgAAdCGMAAAAXQgjAABAF8IIAADQhTACAAB0IYwAAABdCCMAAEAXwggAANCFMAIAAHQhjAAAAF0IIwAAQBfCCAAA0IUwAgAAdCGMAAAAXQgjAABAF8IIAADQhTACAAB0IYwAAABdCCMAAEAXwggAANCFMAIAAHQhjAAAAF0IIwAAQBfCCAAA0IUwAgAAdCGMAAAAXQgjAABAF8IIAADQhTACAAB0IYwAAABdCCMAAEAXwggAANCFMAIAAHQhjAAAAF0IIwAAQBfCCAAA0IUwAgAAdCGMAAAAXQgjAABAF8IIAADQhTACAAB0IYwAAABdCCMAAEAXwggAANCFMAIAAHQhjAAAAF0IIwAAQBfCCAAA0IUwAgAAdCGMAAAAXQgjAABAF8IIAADQhTACAAB0IYwAAABdCCMAAEAXwggAANCFMAIAAHQhjAAAAF0IIwAAQBfCCAAA0IUwAgAAdCGMAAAAXQgjAABAF8IIAADQhTACAAB0IYwAAABdCCMAAEAXwggAANCFMAIAAHQhjAAAAF3MXRipqv9VVX9fVf9aVZur6sqq+nxVvaqq9l2izrqqOnssu7mqLq6q51fVLltp54iqOreqNlXVtVX1mao6eht9O7qqPjuW3zTWP2Ir5XepquPG/iz8LGdX1brlvyMAADCf5i6MJDkuyZ5JPpbkzUn+MsmWJCckubiq7jlZuKoel+S8JIcleV+SU5PsluTkJGcs1kBVHZPkrCT3T/KuJKcnuXuS9VV10hJ1TkqyPsl+Y/l3JfnpJGeNx5suX2P7bxz7c+rYv8OSnDf2GwAA1qxde3dgO9yxtXbD9Maqem2SlyZ5SZL/MW67Y4Zg8KMkh7fWLhi3vyLJOUmOrKqjWmtnTBzngCQnJbkyyaGttY3j9tck+VyS46vqva21T0/UWZfk+CRfTfLg1tpV4/bXJ7kwyUlV9cGFY42OSnJkkg1JHrnwM1XVaUnOT3J6VZ3TWvv+9r9VAABw6zV3V0YWCyKjvx7X95nYdmSSuyQ5YyGITBzj5ePL35k6zjOS7J7k1MnwMAaMPxhf/vZUnYXXr10IImOdjUneOh7v6VN1Ftp9+eTP1Fr7XJJ3j/0+cpGfEwAA1oS5CyNb8avj+uKJbY8Y1x9epPx5Sa5Psq6qdl9mnQ9NldmuOlW1R5J1Y/ufXEE7i6qqCxdbktxvOfUBAKCHebxNK0lSVS9McvskeyU5NMl/zRBEXjdR7L7j+tLp+q21LVV1eZKfSnKvJJcso84VVXVdkv2r6natteuras8k90hybWvtikW6+uVxfdDEtnsn2SXJZa21LcusAwAAa8pMw0hV/Vhr7YezPOZWvDDJj0+8/nCSp7XWvjOxba9xvWmJYyxs33uFdfYcy12/A9uYrrOk1tohi20fr448aDnHAACA1Tbr27S+MT569ydnfNybaa3drbVWSe6W5AkZrm58vqp8+QYAgDkw6zBymyQvSvKlqvpYVT1xa3N5zEJr7d9aa+9L8qgk+yb5i4ndC1cY9rpZxf+8/ertqLNpar0j2rh6if0AADD3Zh1G7p7kNzMMyn5khidcfb2qXjs+MneHaa39S5IvJvmpqrrzuPlL4/pmYy+qatckB2aYo+SyiV1bq7Nfhlu0vt5au35s97ok30hy+3H/tIWne02OQflqhscN32vsx3LqAADAmjLTMNJa+0Fr7a9aa4dneJLTmzKMS3lJkq+Ms4s/rqp21FO87j6ufzSuzxnXj16k7GFJbpdkQ2vtxontW6vzmKky21VnfJTvhrH9h66gHQAAWDN22KN9W2uXttaOz/CkqYWrJY9O8jdJvlZVJ1TV3bd2jGlVdVBV3ezWpqq6zTjp4V0zhIuFuT7OTPLdJEdV1aET5fdIcuL48m1Th3tHkhuTHDN5Naeq9skwqWKSnDZVZ+H1y8ZyC3UOSPLc8XjvmKqz0O6JY38W6jw4yZOSfCfJe6d/VgAAWCt2+KN9W2s/qKq/TXLnDLcf3X1cXpnkJVX1tiQvnro6sZRfTvKHVXV+ksuTfC/DE7UelmEA+7eSPHui7Wuq6tkZQsm5VXVGhpnVH5vhEb5nZphgcLK/l1fVi5KckuSCqnp3kh9kmIBw/yRvmJx9fayzoaremOQFSS6uqjOT7JYhVNwpybFTs68nyRkZBt4fmWHg/VkZxrw8KcNjf5/dWrtmGe8JAADMpR0aRqrq55P8VpLfSLJHkmsyfMl/e4ZHzr4gybEZZiifngl9MX+X5CczzCnywAyPvr0uw9iKdyY5pbV25WSF1tr7q+phSV6W5IljP74ytn1Ka61NN9Jae0tVbczw+OCnZriC9MUMs6X/+WIda60dX1VfyHAl5DlJbkpyUZLXt9Y+uEj5VlVPznC71jPG9+GGDJMxntha27CM9wMAAObWzMNIVd0hyVMyhJD7J6kkn0/yx0n+qrW2eSx6cVW9M8P8IEdmGWGktfZPSY5ZaZ9aa5/KcFVlJXXOSnLWCuusT7J+BeW3JDl5XAAAYKcy60kP/yzDVZDbZRgn8c4kf9xa++xi5VtrP6qqc5M8Ypb9AAAAbv1mfWXk6RkeW3takndM3zK1hHOTvGbG/QAAAG7lZh1GHt1a++hKKoy3UH1qxv0AAABu5WY9z8iKgggAALDzmmkYqapHVtXbl5o/pKruPu4/fJbtAgAA82fWt2kdm+R+rbVvLraztfbNqvqFJHtlGCsCAADspGY9A/uDMsybsTXnJzl0G2UAAIA1btZh5K5JFr0qMuHfxnIAAMBObNZhZFOSe26jzD0zzJoOAADsxGYdRj6b5PFVdbfFdo4D2x8/lgMAAHZisw4jb0lyhySfrKrHVtXuSVJVu1fV45Kcl+T2SU6ZcbsAAMCcmenTtFprH62q30/yiiTvS9Kq6qok+ySpcfn91tqHZ9kuAAAwf2Z9ZSSttVcleXSSs5NcmeExvlcm+dskvzTuBwAAdnKznmckyb/PxG42dgAAYEkzvzICAACwHDvkykiSVNXtMowV2WWx/a21r+2otgEAgFu/mYeRqnpKkhcnOXgrxdqOaBsAAJgfMw0EVfW0JG9P8qMkn0zyr0m2zLINAABgbZj11YkXJrkqyX9trV0y42MDAABryKwHsP9kkvcIIgAAwLbMOoxcmeTGGR8TAABYg2YdRj6Y5PCqqhkfFwAAWGNmHUZekmT3JKdV1e1nfGwAAGANmfUA9vckuT7Js5L8t6r6cpKrFynXWmuPnHHbAADAHJl1GDl84r/3TPKAJcq1GbcLAADMmZmGkdbarG/7AgAA1ijhAQAA6EIYAQAAuph5GKmq21TVsVX1D1W1qaq2TOx7YFX9cVUdNOt2AQCA+TLTMFJVuyX5WJI3Jbl3ku8nmZxz5PIkz0jy32fZLgAAMH9mfWXkRUkenuTVSX48yZ9O7mytXZ3kvCS/NON2AQCAOTPrMPLfk3yqtfaa1tpNWfwRvpcn+YkZtwsAAMyZWYeRA5P8wzbKXJnkTjNuFwAAmDOzDiM3JNl7G2V+IovPyg4AAOxEZh1G/jHJo8aB7DdTVXtlGC/y2Rm3CwAAzJlZh5E/SXLPJH9ZVXec3FFVeydZn2SfJKfNuF0AAGDO7DrLg7XW/p+q+sUkT0vy2CRXJUlVXZDkp5LsnuStrbWzZ9kuAAAwf2Y+6WFr7RkZ5hL5YpK7ZJhn5EFJvpLkma21Y2fdJgAAMH9memVkQWttfZL1VXXbDLdlbWqtXbcj2gIAAObTDgkjC1prm5Ns3pFtAAAA82nmt2kBAAAsx0yvjFTVZcss2lpr955l2wAAwHyZ9W1at0nSFtm+d5K9xv/+ZpIfzrhdAABgzsz60b4HLLWvqn4yySlJ9sww8SEAALATW7UxI621ryR5QpJ7JHnVarULAADcOq3qAPbW2g1JPpbkyavZLgAAcOvT42laW5LcrUO7AADArciqhpGqunOSX0vyr6vZLgAAcOsz60f7vnIr7dwzyeMyPFXrJbNsFwAAmD+zfrTvCdvYf02SE1trfzTjdgEAgDkz6zDy8CW235TkqiT/3FrbMuM2AQCAOTTreUY+McvjAQAAa1ePp2kBAADMfAD7T2xv3dba12bZFwAA4NZt1mNGNiZp21GvZfZ9AQAAbsVmHQD+IskBSQ5LsinJPyb5VoZJDh+Q4bG+n8gQWgAAgJ3YrMPIHyb5dJKTk7y6tXbNwo6qumOSVyd5apLfaq1dOuO2AQCAOTLrAeyvS/KF1trxk0EkSVpr17TWjkvy/43lAACAndisw8hhSc7fRpnzkzxsxu0CAABzZtZhZPcM40O2Zr+xHAAAsBObdRj5fJKjquqBi+2sqkOSPCnJRTNuFwAAmDOzHsD+6iQfTvIPVfWXSc5L8m9JfjzDrVn/LUMAevWM2wUAAObMTMNIa+3vquqoJP93kqclOXpidyW5KslzWmt/P8t2AQCA+TPziQZba2dW1YeSPC7JgzLMLbIpw61ZH2itXTfrNgEAgPmzQ2Y9HwPHX40LAADAzcx6APt/UlX7VNU9d2QbAADAfJp5GKmq21fVG6rqW0m+m+TyiX0/V1VnV9WDZt0uAAAwX2YaRqpqrySfTnJckm8muSTDwPUFX0jy0CRPnmW7AADA/Jn1lZGXJfmpJE9rrT0oyXsmd7bWrk/yiSSPnHG7AADAnJl1GHlCko+01v5iK2X+Jck9ZtwuAAAwZ2YdRvZPcvE2ylyb4XG/AADATmzWYeT7Se66jTIHZhjYDgAA7MRmHUY+l+SIqrrDYjurar8kv5zk/Bm3CwAAzJlZh5E3J9k3ydlVdfDkjvH1e5LskeSU7Tl4Ve1bVc+qqvdV1VeqanNVbaqq86vqmVW16M9TVevGRwpfOda5uKqeX1W7bKWtI6rq3PH411bVZ6rq6G307+iq+uxYftNY/4itlN+lqo4b+7N57N/ZVbVu+e8KAADMp5nOwN5a+0hVvTrJq5L8U5IfJklVfTfJPhke8/vi1tqG7Wzi15O8LckVST6e5GtJfjzDwPk/TfKYqvr11lpbqFBVj0vy3iQ3JHl3kiuT/GqSk5M8ZDzmf1JVxyR5S5LvJXlXkh8kOTLJ+qr66dbaCxepc1KS45N8PcnpSXZLclSSs6rq2NbaqVPlK8kZ43G/lOTUJHdK8qQk51XVE1trH9iO9wgAAObCTMNIkrTWXl1V5yV5XpKfz3ClpCU5O8nJrbVzbsHhL03y2CR/21q7aWFjVb00yWeTPDFDMHnvuP2OGYLBj5Ic3lq7YNz+iiTnJDmyqo5qrZ0xcawDkpyUIbQc2lrbOG5/TYbb0I6vqve21j49UWddhiDy1SQPbq1dNW5/fZILk5xUVR9cONboqAxBZEOSR7bWbhjrnJbhNrbTq+qc1tr3b8H7BQAAt1qznvTwsKp6QGvt4621X2ut7dda2621dpfW2q/ewiCS1to5rbWzJoPIuP1bSU4bXx4+sevIJHdJcsZCEBnL35Dk5ePL35lq5hlJdk9y6mR4GAPGH4wvf3uqzsLr1y4EkbHOxiRvHY/39Kk6C+2+fCGIjHU+l+EKzl3G/gMAwJo06zEjH0/ynBkfc7l+OK63TGx7xLj+8CLlz0tyfZJ1VbX7Mut8aKrMdtWpqj2SrBvb/+QK2gEAgDVj1rdpfTfJ5hkfc5uqatckTx1fTgaC+47rS6frtNa2VNXlGWaMv1eSS5ZR54qqui7J/lV1u9ba9VW1Z4ZJHK9trV2xSPe+PK4Pmth27yS7JLmstbbl5lUWrbOkqrpwiV33W059AADoYdZXRs7N8Bf/1fa6JPdPcnZr7SMT2xcmV9y0RL2F7XtvR529ptY7oo29l9gPAABzb9ZXRl6e5DNV9ftJXtNa++G2KtxSVfW8DIPH/znJU3Z0e7dGrbVDFts+XjF50Cp3BwAAlmXWYeQlGR7p+9Ikz6yq/zfJtzI8TWtSa60985Y2Nj6C981JvpjhiVRXThWZvooxbWH71VN17jzu+95W6myaWq+0jZXWAQCANWXWYeRpE/99t3FZTEtyi8JIVT0/w1wh/5QhiHx7kWJfSnJohrEX/2lcxTjO5MAMA94vm6pz57HOp6fq7JdkzyRfb61dnyStteuq6htJ7lFV+y0ybuQ+43pyDMpXMzxu+F5Vtesi40YWqwMAAGvKrMeMHLjM5V63pJGqenGGIPKPSR6+RBBJhrlEkuTRi+w7LMntkmxord24zDqPmSqzXXXGR/luGNt/6AraAQCANeMWh5GqempV/UyStNb+ZbnLLWjvFRkGrF+Y4YrId7dS/MwMT/g6qqoOnTjGHklOHF++barOO5LcmOSYcQLEhTr7ZLj9LPmPOU0y9fplY7mFOgckee54vHdM1Vlo98SxPwt1HpxhFvbvZJy8EQAA1qJZ3Ka1PskJSS5e2FBVRyc5urU203kyxuO+JsMtTp9M8ryqmi62sbW2Pklaa9dU1bMzhJJzq+qMDDOrPzbDI3zPzDDB4L9rrV1eVS9KckqSC6rq3Ul+kGECwv2TvGFy9vWxzoaqemOSFyS5uKrOTLJbhlBxpyTHTs2+niRnZJgt/sgkn6+qszLMVv+kDI/9fXZr7ZoVv0kAADAnZj1mZMEBSR62A4574LjeJcnzlyjziQwBKUnSWnt/VT0sycuSPDHJHkm+kiE4nNJamx5cn9baW6pqY5IXZpi/5DYZBsm/vLX254s12lo7vqq+kOFKyHOS3JTkoiSvb619cJHyraqenOF2rWckOTbJDRkmYzyxtbZh6bcBAADm344KIztEa+2EDFdhVlrvU0l+eYV1zkpy1grrrM9EEFpG+S0Zxr6cvJJ2AABgLZj1AHYAAIBlEUYAAIAuZhVGbjbuAgAAYGtmNWbkhKo6YXpjVf1oifKttTZX41UAAIDZmlUguNnzdWdcHgAAWGNucRhprRl3AgAArJggAQAAdCGMAAAAXQgjAABAF8IIAADQhTACAAB0IYwAAABdCCMAAEAXwggAANCFMAIAAHQhjAAAAF0IIwAAQBfCCAAA0IUwAgAAdCGMAAAAXQgjAABAF8IIAADQhTACAAB0IYwAAABdCCMAAEAXwggAANCFMAIAAHQhjAAAAF0IIwAAQBfCCAAA0IUwAgAAdCGMAAAAXQgjAABAF8IIAADQhTACAAB0IYwAAABdCCMAAEAXwggAANCFMAIAAHQhjAAAAF0IIwAAQBfCCAAA0IUwAgAAdCGMAAAAXQgjAABAF8IIAADQhTACAAB0IYwAAABdCCMAAEAXwggAANCFMAIAAHQhjAAAAF0IIwAAQBfCCAAA0IUwAgAAdCGMAAAAXQgjAABAF8IIAADQhTACAAB0IYwAAABdCCMAAEAXwggAANCFMAIAAHQhjAAAAF0IIwAAQBfCCAAA0IUwAgAAdCGMAAAAXQgjAABAF8IIAADQhTACAAB0IYwAAABdCCMAAEAXwggAANDF3IWRqjqyqt5SVZ+sqmuqqlXVu7ZRZ11VnV1VV1bV5qq6uKqeX1W7bKXOEVV1blVtqqprq+ozVXX0Nto5uqo+O5bfNNY/Yivld6mq48b+bB77d3ZVrdv2OwEAAPNt7sJIkpcnOSbJA5J8Y1uFq+pxSc5LcliS9yU5NcluSU5OcsYSdY5JclaS+yd5V5LTk9w9yfqqOmmJOiclWZ9kv7H8u5L8dJKzxuNNl6+x/TeO/Tl17N9hSc4b+w0AAGvWPIaR45IclOSOSX5nawWr6o4ZgsGPkhzeWntma+1FGYLMp5McWVVHTdU5IMlJSa5Mcmhr7bmtteOS/EySryY5vqp+YarOuiTHj/t/prV2XGvtuUkOGY9z0njcSUclOTLJhiQPaK29qLX2zCQPH/t7elXdYblvCgAAzJu5CyOttY+31r7cWmvLKH5kkrskOaO1dsHEMW7IcIUluXmgeUaS3ZOc2lrbOFHnqiR/ML787ak6C69fO5ZbqLMxyVvH4z19qs5Cuy8f+7NQ53NJ3j32+8ht/oQAADCn5i6MrNAjxvWHF9l3XpLrk6yrqt2XWedDU2W2q05V7ZFk3dj+J1fQDgAArBm79u7ADnbfcX3p9I7W2paqujzJTyW5V5JLllHniqq6Lsn+VXW71tr1VbVnknskuba1dsUiffjyuD5oYtu9k+yS5LLW2pZl1llSVV24xK77Lac+AAD0sNavjOw1rjctsX9h+97bUWevqfWOaGPvJfYDAMDcW+tXRnYKrbVDFts+XjF50Cp3BwAAlmWtXxmZvooxbWH71dtRZ9PUeke0cfUS+wEAYO6t9TDypXF9s7EXVbVrkgOTbEly2TLr7JdkzyRfb61dnySttesyzHdy+3H/tPuM68kxKF/N8Pjee439WBjgZrMAAA5qSURBVE4dAABYU9Z6GDlnXD96kX2HJbldkg2ttRuXWecxU2W2q874KN8NY/sPXUE7AACwZqz1MHJmku8mOaqqDl3YOD5a98Tx5dum6rwjyY1JjpmcqLCq9kny0vHlaVN1Fl6/bCy3UOeAJM8dj/eOqToL7Z449mehzoOTPCnJd5K8dxs/HwAAzK25G8BeVY9P8vjx5d3G9S9U1frxv7/bWnthkrTWrqmqZ2cIJedW1RkZZkR/bIZH+J6ZYYLBf9dau7yqXpTklCQXVNW7k/wgwwSE+yd5Q2vt01N1NlTVG5O8IMnFVXVmkt0yhIo7JTl2cgLF0RlJnjAe9/NVdVaSfcc6uyR5dmvtmu14iwAAYC7MXRhJ8oAkR09tu9e4JMm/JHnhwo7W2vur6mFJXpbkiUn2SPKVDMHhlMVmcm+tvaWqNo7HeWqGK0hfzDBb+p8v1qnW2vFV9YUMV0Kek+SmJBcleX1r7YOLlG9V9eQMt2s9I8mxSW7IMBnjia21Ddt+KwAAYH7NXRhprZ2Q5IQV1vlUkl9eYZ2zkpy1wjrrk6xfQfktSU4eFwAA2Kms9TEjAADArZQwAgAAdCGMAAAAXQgjAABAF8IIAADQhTACAAB0IYwAAABdCCMAAEAXwggAANCFMAIAAHQhjAAAAF0IIwAAQBfCCAAA0IUwAgAAdCGMAAAAXQgjAABAF8IIAADQhTACAAB0IYwAAABdCCMAAEAXwggAANCFMAIAAHQhjAAAAF0IIwAAQBfCCAAA0IUwAgAAdCGMAAAAXQgjAABAF8IIAADQhTACAAB0IYwAAABdCCMAAEAXwggAANCFMAIAAHQhjAAAAF0IIwAAQBfCCAAA0IUwAgAAdCGMAAAAXQgjAABAF8IIAADQhTACAAB0IYwAAABdCCMAAEAXwggAANCFMAIAAHQhjAAAAF0IIwAAQBfCCAAA0IUwAgAAdCGMAAAAXQgjAABAF8IIAADQhTACAAB0IYwAAABdCCMAAEAXwggAANCFMAIAAHQhjAAAAF0IIwAAQBfCCAAA0IUwAgAAdCGMAAAAXQgjAABAF8IIAADQhTACAAB0IYwAAABdCCMAAEAXwggAANCFMAIAAHQhjAAAAF0IIwAAQBfCCAAA0IUwAgAAdCGMdFZV+1fV26vqm1V1Y1VtrKo3VdU+vfsGAAA70q69O7Azq6p7J9mQ5K5JPpDkn5P8bJLfTfLoqnpIa+17HbsIAAA7jCsjff1xhiDyvNba41trv9dae0SSk5PcN8lru/YOAAB2IGGkk/GqyKOSbEzy1qndr0pyXZKnVNWeq9w1AABYFW7T6ufh4/qjrbWbJne01r5fVZ/KEFZ+Psnfr3bnbqkDfu9ve3dh1W183a/07gIAwFwRRvq577i+dIn9X84QRg7KNsJIVV24xK7/85JLLskhhxyyfT28Ba74xqZVb7O33df/bu8urLr732Ov3l0AgCXtjN9HDvnYK1e9zUsuuSRJDtieusJIPwvf4pb6LVnYvvctaONHmzdv3nTRRRdtXEbZ+43rf74F7TF/btF5v+jfZtgTVovf9Z2T875zct53Pvcb/21e7XN+QJJrtqeiMLIGtNZu8aWPhasrszgW88N53/k45zsn533n5LzvfObxnBvA3s/ClY+l7nNZ2H71KvQFAABWnTDSz5fG9UFL7L/PuF5qTAkAAMw1YaSfj4/rR1XVfzoPVXWHJA9Jcn2Sf1jtjgEAwGoQRjpprX01yUczDPh57tTuVyfZM8k7W2vXrXLXAABgVRjA3tf/SLIhySlV9cgklyT5uQxzkFya5GUd+wYAADtUtdZ692GnVlX3TPKaJI9Osm+SK5K8L8mrW2tX9ewbAADsSMIIAADQhTEjAABAF8IIAADQhTACAAB0IYwAAABdCCMAAEAXwggAANCFMLKTq6r9q+rtVfXNqrqxqjZW1Zuqap/efWPrqmrfqnpWVb2vqr5SVZuralNVnV9Vz6yqRX+/q2pdVZ1dVVeOdS6uqudX1S5baeuIqjp3PP61VfWZqjp6x/10rERV/WZVtXF51hJlVnwOq+roqvrsWH7TWP+IHfNTsBxV9cjxd/5b42f2N6vqI1X1y4uU9bu+BlTVr1TVR6vq6+N5vKyq3lNVv7BEeed9DlTVkVX1lqr6ZFVdM35+v2sbdVbl3K76Z39rzbKTLknuneTfkrQk70/yuiTnjK//Ocm+vfto2er5++3xXH0zyV8m+cMkb09y9bj9zIxzCU3UeVySLUmuTfJnSV4/nuuW5D1LtHPMuP+7Sd6a5OQk/zpuO6n3+7CzL0nuOZ7z74/n5FmzOIdJThr3/+tY/q1JvjduO6b3z70zLkn+aOKc/EmSP0hyepKLkvzRVFm/62tgSfK/Js7Jn47/Tp+Z5AdJbkrym877fC5J/nF8j7+f5JLxv9+1lfKrcm57fPZ3PxmWfkuSj4z/cx07tf2N4/bTevfRstXz94gkv5rkNlPb75bka+M5fOLE9jsm+XaSG5McOrF9jyQbxvJHTR3rgCQ3jB9EB0xs3yfJV8Y6v9D7vdhZlySV5O+SfHX8h+lmYWR7zmGSdeP2ryTZZ+pY3xuPd8CO+rksi57rZ4/nZH2S3RbZ/2MT/+13fQ0s42f5j5J8K8ldp/Y9fDwnlznv87mM5/A+4+f44dlKGFmtc9vrs99tWjupqrp3kkcl2Zgh9U56VZLrkjylqvZc5a6xTK21c1prZ7XWbpra/q0kp40vD5/YdWSSuyQ5o7V2wUT5G5K8fHz5O1PNPCPJ7klOba1tnKhzVYa/yibDFRr6eF6GUPr0DL+zi9mec7jw+rVjuYU6GzN8Xuw+tskqqKrdk7w2wx8ZntNa+8F0mdbaDyde+l1fG/6PDLfTf6a19u3JHa21j2f4i/pdJjY773Oktfbx1tqX2/htfxtW69x2+ewXRnZeDx/XH13ky+z3k3wqye2S/Pxqd4yZWPhismVi2yPG9YcXKX9ekuuTrBu/+CynzoemyrCKqurgDLdsvLm1dt5Wim7POXTeb11+McMXkb9JctM4huDFVfW7S4wb8Lu+Nnw5w+1YP1tVd57cUVWHJblDhiujC5z3tWu1zm2X/x+EkZ3Xfcf1pUvs//K4PmgV+sIMVdWuSZ46vpz8QFnynLfWtiS5PMmuSe61zDpXZPhr/P5Vdbtb2G1WYDzH78zwl/KXbqP4is7heDX0HkmuHfdP89mw+h48rm9I8vkkH8wQRN+UZENVfaKqJv9C7nd9DWitXZnkxUl+PMkXq+pPquoPq+qvk3w0yceS/NZEFed97drh57bnZ78wsvPaa1xvWmL/wva9V6EvzNbrktw/ydmttY9MbN+ec77cOnstsZ8d45VJHpjkaa21zdsou9Jz6LPh1ueu4/pFGe7nfmiGv4r/TIYvpYclec9Eeb/ra0Rr7U1JnpDhi+azk/xekl/PMLh4/dTtW8772rUa57bbZ78wAmtIVT0vyfEZnrDxlM7dYQeoqp/LcDXkDa21T/fuD6ti4d/qLUke21o7v7V2bWvtC0l+LcnXkzxsqUe9Mr+q6n9meHrW+gxPwNwzySFJLkvyl1X1R/16B7MhjOy8tvXXjoXtV69CX5iBqjomyZuTfDHJw8dL/JO255wvt85Sf0lhhsbbs/4iw2X3Vyyz2krPoc+GW5+F9/rzkwNRk6S1dn2GJyMmyc+Oa7/ra0BVHZ7h0b7/u7X2gtbaZa2161trF2UIod9IcnxVLdya47yvXatxbrt99gsjO68vjeul7v27z7heakwJtyJV9fwkb0nyTxmCyLcWKbbkOR+/5B6Y4S+vly2zzn4Z/kr39fELETve7TOci4OT3DAx0WHL8BS8JDl93Pam8fWKzmFr7boMX3JuP+6f5rNh9S2cw6W+BCw89ea2U+X9rs+3hUnmPj69YzwPn83wPe6B42bnfe3a4ee252e/MLLzWvhwe1RNzdRdVXdI8pAMT2f4h9XuGCtTVS/OMDHRP2YIIt9eoug54/rRi+w7LMPT0za01m5cZp3HTJVhx7sxw2RXiy2fH8ucP75euIVre86h837r8vcZxor8l+nP69H9x/Xl49rv+tqw8GSkuyyxf2H7wqOenfe1a7XObZ//H2Y9cYllfpaY9HDulwy36rQkFyS50zbK3jHJd7KySZMOjAmx5mJJckIWn/RwxecwJj281S1JPjCek+Omtj8qw0zcVyXZa9zmd30NLEl+Y3zfv5XkHlP7HjOe981J9nXe53vJ8iY93OHnttdnf42NsBMaJz7ckOFJLR9IckmSn8swB8mlSda11r7Xr4dsTVUdnWFQ448y3KK12D29G1tr6yfqPD7DYMgbkpyR5Mokj83wCMAzk/xGm/pQqKpjk5yS4YPo3Rn+Cndkkv0zDKJ+4Sx/LrZPVZ2Q4VatZ7fW/nRq34rPYVW9IckLMgyOPjPJbkmelGTfDH/AOHWH/TDcTFXtn+Hz+p4ZrpR8PsOXjcfnP76IvHeivN/1OTdeBftIkv8rwwSH78sQTA7OcAtXJXl+a+3NE3Wc9zkxnqvHjy/vluSXMtxm9clx23cn3/vVOrddPvt7p0FL3yXDP2zvSHLF+D/pv2R4dv0+vftm2ea5OyHDl5CtLecuUu8hSc7O8JfUzUm+kOS4JLtspa1fTfKJDP8gXpfkc0mO7v0eWBb9/+FZS+xf8TlM8rSx3HVjvU8kOaL3z7qzLhluy3nL+Dn9gyTfzfAF9WeXKO93fc6XJD+W5PkZbpm+JsO4gG9nmGvmUc77/C7L+Dd8Y69zu9qf/a6MAAAAXRjADgAAdCGMAAAAXQgjAABAF8IIAADQhTACAAB0IYwAAABdCCMAAEAXwggAANCFMAIAAHQhjAAAAF0IIwAAQBfCCAAA0IUwAgAAdCGMAAAAXQgjAABAF8IIAADQhTACAAB08f8DcKINREnfiJMAAAAASUVORK5CYII=\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "image/png": {
       "height": 248,
       "width": 401
      },
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "loci_df.QUAL.plot(kind=\"hist\");"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAyMAAAHwCAYAAAC15rBtAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAAWJQAAFiUBSVIk8AAAADh0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uMy4yLjEsIGh0dHA6Ly9tYXRwbG90bGliLm9yZy+j8jraAAAgAElEQVR4nO3de7hlVXkn6t8nRBBUQNSIYp5SI0qruQB2EkwQ5cRomqhR0mJ3FMVLTAuJBD2JtwgG054Wb4iRPqaljCQHE2xNl0FjIiJoGS9gGj2ieKE0KhoVKIQCDDj6jzm3WVnsXbV31ao9WLve93nWM11zjm+OsfdaLvav5hpzVGstAAAAq+0OvQcAAADsmoQRAACgC2EEAADoQhgBAAC6EEYAAIAuhBEAAKALYQQAAOhCGAEAALoQRgAAgC6EEQAAoAthBAAA6EIYAQAAuti99wDYearqyiR3TbKp81AAAFi71iW5rrV2v5UWCiNr213vdKc73e3ggw++W++BAACwNl1++eW58cYbt6tWGFnbNh188MF3u+SSS3qPAwCANerQQw/NpZdeuml7as0ZAQAAuhBGAACALoQRAACgC2EEAADoQhgBAAC6EEYAAIAuhBEAAKALYQQAAOhCGAEAALoQRgAAgC6EEQAAoAthBAAA6EIYAQAAuhBGAACALoQRAACgC2EEAADoQhgBAAC6EEYAAIAudu89ANamdX/wN72HsOo2vfo/9B4CAMBccWUEAADoQhgBAAC6EEYAAIAuhBEAAKALYQQAAOhCGAEAALoQRgAAgC6EEQAAoAthBAAA6EIYAQAAuhBGAACALoQRAACgC2EEAADoQhgBAAC6EEYAAIAuhBEAAKALYQQAAOhCGAEAALoQRgAAgC6EEQAAoAthBAAA6EIYAQAAuhBGAACALoQRAACgC2EEAADoQhgBAAC6EEYAAIAuhBEAAKALYQQAAOhCGAEAALoQRgAAgC6EEQAAoAthBAAA6EIYAQAAuhBGAACALoQRAACgC2EEAADoQhgBAAC6EEYAAIAuhBEAAKALYQQAAOhCGAEAALoQRgAAgC6EEQAAoAthBAAA6EIYAQAAuhBGAACALoQRAACgC2EEAADoQhgBAAC6EEYAAIAuhBEAAKALYQQAAOhCGAEAALqY+zBSVb9ZVW18PHuJNkdX1YVVtbmqrq+qj1fVcds473FV9Ymx/eax/uittN+tqk6qqsuq6saqurqqzq+qw7dSc6eqOrWqvlBVN1XVP1fVX1bVwcv/DQAAwHya6zBSVfdNcmaS67fS5oQkG5I8NMk5Sd6a5N5J1lfV6UvUnJ5kfZIDxvbnJHlYkg3j+abbV5Jzk7wuyR3HMb07yRFJLqqqJyxSs0eSv0vyh0muS/LGJH+f5NeTfKqqfm6bvwAAAJhju/cewPYaA8DZSb6X5H8meeEibdYlOT3J1UkOa61tGve/Msknk5xcVe9qrX1soubwJCcn+XKSh7fWrhn3vybJJUlOr6r3LpxrdGySY5JsTHJUa+2mseasJB9J8taquqC19v2Jmt9L8ogk5yV5Smvth2PNO5O8J8nbquphC/sBAGCtmecrI7+T5NFJnpnkhiXaHJ9kjyRnToaHMWD88fj0eVM1C89ftRBExppNSd48nu+ZUzW/PW5fthBExppPJnlnkntkCCtJfhSkFvr5vycDR2vtr5NcnOTfJXnkEj8XAADMvbkMI+OcilcneWNr7aKtNH30uH3/IsfeN9Vmu2qqas8khyfZkiFELKefByT5iSRXtNauXMHYAABgzZi7r2lV1e5J3pHka0leso3mDxq3V0wfaK1dVVU3JDmwqvZqrW2pqr2T3CfJ9a21qxY53xfH7UET+x6QZLckX2mt3bLMmiXHtZWaJVXVJUscevBy6gEAoIe5CyMZJnz/bJJfbK3duI22+4zbzUsc35xk77HdlmW2T5J9V9jHLGoAAGBNmaswMt5h6iVJXjs56XxX11o7dLH94xWTQ1Z5OAAAsCxzM2dk/HrWn2X4atPLl1m2cIVhnyWOT1+hWG77a7ejjx2tAQCANWVuwkiSO2eYQ3FwkpsmFjpsSV4xtnnruO8N4/MvjNvbzL2oqgMyfEXr6621LUnSWrshyTeS3Hk8Pu2B43ZyrseXk9ya5P5jYFpOzZLj2koNAACsKfP0Na2bk/yPJY4dkmEeyUcy/KG/8BWuCzKs5fHYiX0LHjfRZtIFSZ421py9rZrW2k1VtTHJL42PDy2jny9nmIB/UFXdb5E7ai01NgAAWDPm5spIa+3G1tqzF3sk+V9js7eP+945Pj87Q4g5YVwAMUlSVfvlX+/EddZUVwvPXzq2W6hZl+T54/mmQ8pbxu1p461+F2oenuQpSb6T5F0TP0ub6Oe/VdUdJmqekCHUfC7Jh5f+jQAAwHybpysjK9Zau7KqXpTkjCSfGlc3/0GGBQgPzCIT4VtrG6vqdRlWSL+sqs5LcscMoeJuSU6cWn09Sc5N8qTxvJ+uqg1J9h9rdkvynNbadVM1r0ty9Fjz8ar6YIa1R34jw529jrf6OgAAa9maDiNJ0lp7U1VtSvLCJE/PcDXocxlWS3/7EjUnV9VnMlwJeW6SHya5NMlrWmvvXaR9q6qnJtmYYdX3E5PclOSiJKe11jYuUnNzVf1ykj9I8tQkJyW5Lsl7kryitfa5HfrBAQDgdm5NhJHW2ilJTtnK8Q1JNqzwnOuTrF9B+1uSvH58LLdmS4Z1U/5wJWMDAIC1YG7mjAAAAGuLMAIAAHQhjAAAAF0IIwAAQBfCCAAA0IUwAgAAdCGMAAAAXQgjAABAF8IIAADQhTACAAB0IYwAAABdCCMAAEAXwggAANCFMAIAAHQhjAAAAF0IIwAAQBfCCAAA0IUwAgAAdCGMAAAAXQgjAABAF8IIAADQhTACAAB0IYwAAABdCCMAAEAXwggAANCFMAIAAHQhjAAAAF0IIwAAQBfCCAAA0IUwAgAAdCGMAAAAXQgjAABAF8IIAADQhTACAAB0IYwAAABdCCMAAEAXwggAANCFMAIAAHQhjAAAAF0IIwAAQBfCCAAA0IUwAgAAdCGMAAAAXQgjAABAF8IIAADQhTACAAB0IYwAAABdCCMAAEAXwggAANCFMAIAAHQhjAAAAF0IIwAAQBfCCAAA0IUwAgAAdCGMAAAAXQgjAABAF8IIAADQhTACAAB0IYwAAABdCCMAAEAXwggAANCFMAIAAHQhjAAAAF0IIwAAQBfCCAAA0IUwAgAAdCGMAAAAXQgjAABAF8IIAADQhTACAAB0IYwAAABdCCMAAEAXwggAANDF3IWRqvp/quqDVfVPVXVjVV1dVZ+uqldU1f5L1BxeVeePbW+sqsuq6gVVtdtW+jm6qi6sqs1VdX1VfbyqjtvG2I6rqk+M7TeP9Udvpf1uVXXSOJ6Fn+X8qjp8+b8RAACYT3MXRpKclGTvJH+X5I1J/jzJLUlOSXJZVd13snFVPSHJRUmOSPLuJGcmuWOS1yc5d7EOquqEJBuSPDTJOUnemuTeSdZX1elL1JyeZH2SA8b25yR5WJIN4/mm29fY/+vG8Zw5ju+IJBeN4wYAgDVr994D2A53ba3dNL2zql6V5CVJXpzkv4z77pohGNya5MjW2qfG/S9PckGSY6rq2NbauRPnWZfk9CRXJzmstbZp3P/KJJ9McnJVvau19rGJmsOTnJzky0ke3lq7Ztz/miSXJDm9qt67cK7RsUmOSbIxyVELP1NVnZXkI0neWlUXtNa+v/2/KgAAuP2auysjiwWR0V+O2wdO7DsmyT2SnLsQRCbO8bLx6W9Pnef4JHskOXMyPIwB44/Hp8+bqll4/qqFIDLWbEry5vF8z5yqWej3ZZM/U2vtk0neOY77mEV+TgAAWBPmLoxsxa+N28sm9j163L5/kfYXJdmS5PCq2mOZNe+barNdNVW1Z5LDx/4vXkE/AACwZszj17SSJFX1wiR3TrJPksOS/GKGIPLqiWYPGrdXTNe31m6pqiuTPCTJ/ZNcvoyaq6rqhiQHVtVerbUtVbV3kvskub61dtUiQ/3iuD1oYt8DkuyW5CuttVuWWbOkqrpkiUMPXk49AAD0MNMwUlU/1lr7l1mecytemOTHJ56/P8kzWmvfmdi3z7jdvMQ5Fvbvu8Kavcd2W3ZiH9M1AACwpsz6a1rfGG+9+5MzPu9ttNbu1VqrJPdK8qQMVzc+XVWH7Oy+b29aa4cu9kjy+d5jAwCApcw6jNwhyYuSfKGq/q6qnry1tTxmobX27dbau5M8Jsn+Sf5s4vDCFYZ9blP4b/dfux01m6e2O6OPa5c4DgAAc2/WYeTeSX4zw6TsozLc4errVfWq8Za5O01r7atJPpfkIVV193H3F8btbeZeVNXuSe6XYY2Sr0wc2lrNARm+ovX11tqWsd8bknwjyZ3H49MW7u41OQflyxluN3z/cRzLqQEAgDVlpmGktfaD1tpftNaOzDB5+g0Z5qW8OMmXxtXFn1BVO+suXvcet7eO2wvG7WMXaXtEkr2SbGyt3Tyxf2s1j5tqs1014618N479/9IK+gEAgDVjp93at7V2RWvt5Ax3mlq4WvLYJP8zydeq6pSquvfWzjGtqg6qqtt8tamq7jAuenjPDOFiYa2P85J8N8mxVXXYRPs9k5w2Pn3L1OnOTnJzkhMmr+ZU1X4ZFlVMkrOmahaev3Rst1CzLsnzx/OdPVWz0O9p43gWah6e5ClJvpPkXdM/KwAArBU7/da+rbUfVNXfJLl7hq8f3Xt8/GGSF1fVW5L8/tTViaX8apL/WlUfSXJlku9luKPWIzNMYP9WkudM9H1dVT0nQyi5sKrOzbCy+uMz3ML3vAwLDE6O98qqelGSM5J8qqremeQHGRYgPDDJaydXXx9rNlbV65L8XpLLquq8JHfMECruluTEqdXXk+TcDBPvj8kw8X5DhjkvT8lw29/ntNauW8bvBAAA5tJODSNV9fNJfivJf0yyZ5LrMvyR/7Ykh2T44/3EDCuUT6+Evpi/T/KTGdYU+dkMt769IcPcinckOaO1dvVkQWvtPVX1yCQvTfLkcRxfGvs+o7XWpjtprb2pqjZluH3w0zNcQfpchtXS377YwFprJ1fVZzJcCXlukh8muTTJa1pr712kfauqp2b4utbx4+/hpgyLMZ7WWtu4jN8HAADMrZmHkaq6S5KnZQghD01SST6d5E+S/EVr7cax6WVV9Y4M64Mck2WEkdbaZ5OcsNIxtdY+muGqykpqNiTZsMKa9UnWr6D9LUlePz4AAGCXMutFD/9Hhqsge2WYJ/GOJH/SWvvEYu1ba7dW1YVJHj3LcQAAALd/s74y8swMt609K8nZ01+ZWsKFSV4543EAAAC3c7MOI49trX1gJQXjV6g+OuNxAAAAt3OzXmdkRUEEAADYdc00jFTVUVX1tqXWD6mqe4/Hj5xlvwAAwPyZ9de0Tkzy4NbaNxc72Fr7ZlX9QpJ9MswVAQAAdlGzXoH9kAzrZmzNR5Icto02AADAGjfrMHLPJIteFZnw7bEdAACwC5t1GNmc5L7baHPfDKumAwAAu7BZh5FPJHliVd1rsYPjxPYnju0AAIBd2KzDyJuS3CXJxVX1+KraI0mqao+qekKSi5LcOckZM+4XAACYMzO9m1Zr7QNV9UdJXp7k3UlaVV2TZL8kNT7+qLX2/ln2CwAAzJ9ZXxlJa+0VSR6b5PwkV2e4je/VSf4mya+MxwEAgF3crNcZSfKjlditxg4AACxp5ldGAAAAlmOnXBlJkqraK8Nckd0WO95a+9rO6hsAALj9m3kYqaqnJfn9JAdvpVnbGX0DAADzY6aBoKqekeRtSW5NcnGSf0pyyyz7AAAA1oZZX514YZJrkvxia+3yGZ8bAABYQ2Y9gf0nk/yVIAIAAGzLrMPI1UlunvE5AQCANWjWYeS9SY6sqprxeQEAgDVm1mHkxUn2SHJWVd15xucGAADWkFlPYP+rJFuSPDvJf6qqLya5dpF2rbV21Iz7BgAA5sisw8iRE/977yQ/s0S7NuN+AQCAOTPTMNJam/XXvgAAgDVKeAAAALoQRgAAgC5mHkaq6g5VdWJV/UNVba6qWyaO/WxV/UlVHTTrfgEAgPky0zBSVXdM8ndJ3pDkAUm+n2RyzZErkxyf5D/Psl8AAGD+zPrKyIuSPCrJqUl+PMmfTh5srV2b5KIkvzLjfgEAgDkz6zDyn5N8tLX2ytbaD7P4LXyvTPITM+4XAACYM7MOI/dL8g/baHN1krvNuF8AAGDOzDqM3JRk3220+Yksvio7AACwC5l1GPnHJI8ZJ7LfRlXtk2G+yCdm3C8AADBnZh1G/t8k903y51V118kDVbVvkvVJ9kty1oz7BQAA5szuszxZa+3/q6pfTvKMJI9Pck2SVNWnkjwkyR5J3txaO3+W/QIAAPNn5osettaOz7CWyOeS3CPDOiOHJPlSkme11k6cdZ8AAMD8memVkQWttfVJ1lfVnTJ8LWtza+2GndEXAAAwn3ZKGFnQWrsxyY07sw8AAGA+zfxrWgAAAMsx0ysjVfWVZTZtrbUHzLJvAABgvsz6a1p3SNIW2b9vkn3G//3NJP8y434BAIA5M+tb+65b6lhV/WSSM5LsnWHhQwAAYBe2anNGWmtfSvKkJPdJ8orV6hcAALh9WtUJ7K21m5L8XZKnrma/AADA7U+Pu2ndkuReHfoFAABuR1Y1jFTV3ZP8epJ/Ws1+AQCA259Z39r3D7fSz32TPCHDXbVePMt+AQCA+TPrW/ueso3j1yU5rbX232bcLwAAMGdmHUYetcT+Hya5JsnnW2u3zLhPAABgDs16nZEPz/J8AADA2tXjbloAAAAzn8D+E9tb21r72izHAgAA3L7Nes7IpiRtO+paZj8WAADgdmzWAeDPkqxLckSSzUn+Mcm3Mixy+DMZbuv74QyhBQAA2IXNOoz81yQfS/L6JKe21q5bOFBVd01yapKnJ/mt1toVM+4bAACYI7OewP7qJJ9prZ08GUSSpLV2XWvtpCT//9gOAADYhc06jByR5CPbaPORJI+ccb8AAMCcmXUY2SPD/JCtOWBsBwAA7MJmHUY+neTYqvrZxQ5W1aFJnpLk0hn3CwAAzJlZT2A/Ncn7k/xDVf15kouSfDvJj2f4atZ/yhCATp1xvwAAwJyZaRhprf19VR2b5L8neUaS4yYOV5Jrkjy3tfbBWfYLAADMn5kvNNhaO6+q3pfkCUkOybC2yOYMX83669baDbPuEwAAmD87ZdXzMXD8xfgAAAC4jVlPYP83qmq/qrrvzuwDAACYTzMPI1V156p6bVV9K8l3k1w5ceznqur8qjpk1v0CAADzZaZhpKr2SfKxJCcl+WaSyzNMXF/wmSS/lOSps+wXAACYP7O+MvLSJA9J8ozW2iFJ/mryYGttS5IPJzlqxv0CAABzZtZh5ElJ/ra19mdbafPVJPeZcb8AAMCcmXUYOTDJZdtoc32G2/0CAAC7sFmHke8nuec22twvw8R2AABgFzbrMPLJJEdX1V0WO1hVByT51SQfmXG/AADAnJl1GHljkv2TnF9VB08eGJ//VZI9k5wx434BAIA5M9Mw0lr72ySnJnlEks8meXGSVNV3x+eHJ3lxa23j9py/qvavqmdX1bur6ktVdWNVba6qj1TVs6pq0Z+nqg4f1ze5eqy5rKpeUFW7baWvo6vqwvH811fVx6vquG2M77iq+sTYfvNYf/RW2u9WVSeN47lxHN/5VXX48n8rAAAwn2a+6GFr7dQMt+79X0muSXJrkpbk/CT/V2vtNTtw+t9I8tYkP5fk40nekORdSR6a5E+T/GVVTa5rkqp6QpKLkhyR5N1JzkxyxySvT3LuYp1U1QlJNoznPWfs895J1lfV6UvUnJ5kfZIDxvbnJHlYkg3j+abb19j/68bxnDmO74gkF43jBgCANWv3WZ6sqo5Icl1r7UNJPjTLc4+uSPL4JH/TWvvhRL8vSfKJJE/OcHvhd43775ohGNya5MjW2qfG/S9PckGSY6rq2NbauRPnWpfk9CRXJzmstbZp3P/KDHNiTq6qd7XWPjZRc3iSk5N8OcnDW2vXjPtfk+SSJKdX1XsXzjU6NskxSTYmOaq1dtNYc1aGOTVvraoLWmvf38HfGQAA3C7N+srIh5I8d8bn/JHW2gWttQ2TQWTc/60kZ41Pj5w4dEySeyQ5dyGIjO1vSvKy8elvT3VzfJI9kpw5GR7GgPHH49PnTdUsPH/VQhAZazYlefN4vmdO1Sz0+7KFIDLWfDLJO8dxHxMAAFijZh1Gvpvkxhmfc7n+ZdzeMrHv0eP2/Yu0vyjJliSHV9Uey6x531Sb7aqpqj0zzJ/ZkuTiFfQDAABrxky/ppXkwgx/ZK+qqto9ydPHp5OB4EHj9orpmtbaLVV1ZZKHJLl/ksuXUXNVVd2Q5MCq2qu1tqWq9s6wovz1rbWrFhneF8ftQRP7HpBktyRfaa3dctuSRWuWVFWXLHHowcupBwCAHmZ9ZeRlSR5UVX9UVT8243NvzaszTDY/f7yj14KFld43L1G3sH/f7ajZZ2q7M/rYd4njAAAw92Z9ZeTFGW7h+5Ikz6qq/53kWxnupjWptdaeNYsOq+p3Mkwe/3ySp83inPOmtXboYvvHKyaHrPJwAABgWWYdRp4x8b/vNT4W05LscBgZb5n7xiSfy3BHqqunmkxfxZi2sP/aqZq7j8e+t5WazVPblfax0hoAAFhTZh1G7jfj8y2pql6QYa2Qz2YIIv+8SLMvJDksw9yLfzOvYpxncr8ME96/MlVz97HmY1M1ByTZO8nXW2tbkqS1dkNVfSPJfarqgEXmjTxw3E7OQflyhtsN37+qdl9k3shiNQAAsKbs8JyRqnp6Vf1UkrTWvrrcxw72+fsZgsg/JnnUEkEkGdYSSZLHLnLsiCR7JdnYWrt5mTWPm2qzXTXjrXw3jv3/0gr6AQCANWMWE9jXJ3ni5I6qOq6qdsof0uOCha/OcKXjqNbad7fS/LwMtxs+tqoOmzjHnklOG5++Zarm7CQ3JzlhXABxoWa/DHNhkn9d0yRTz186tluoWZfk+eP5zp6qWej3tHE8CzUPT/KUJN/JuHgjAACsRbP+mtaCdUkeOeuTVtVxSV6Z4StOFyf5naqabraptbY+SVpr11XVczKEkgur6twMK6s/PsMtfM/LsMDgj7TWrqyqFyU5I8mnquqdSX6QYQHCA5O8dnL19bFmY1W9LsnvJbmsqs5LcscMoeJuSU6cWn09Sc7NsFr8MUk+XVUbkuw/1uyW5DmttetW/EsCAIA5sbPCyM6yMCdltyQvWKLNhzNcrUmStNbeU1WPTPLSJE9OsmeSL2UIDme01qbv9JXW2puqalOSF2ZYv+QOGSbJv6y19vbFOm2tnVxVn8lwJeS5SX6Y5NIkr2mtvXeR9q2qnprh61rHJzkxyU0ZFmM8rbW2celfAwAAzL+5CiOttVOSnLIddR9N8qsrrNmQZMMKa9ZnIggto/0tGea+vH4l/QAAwFow60UPAQAAlmVWYeQ2X3UCAADYmll9TeuUqjplemdV3bpE+9Zam6uviAEAALM1q0Bwm1tazbg9AACwxuxwGGmtmXcCAACsmCABAAB0IYwAAABdCCMAAEAXwggAANCFMAIAAHQhjAAAAF0IIwAAQBfCCAAA0IUwAgAAdCGMAAAAXQgjAABAF8IIAADQhTACAAB0IYwAAABdCCMAAEAXwggAANCFMAIAAHQhjAAAAF0IIwAAQBfCCAAA0IUwAgAAdCGMAAAAXQgjAABAF8IIAADQhTACAAB0IYwAAABdCCMAAEAXwggAANCFMAIAAHQhjAAAAF0IIwAAQBfCCAAA0IUwAgAAdCGMAAAAXQgjAABAF8IIAADQhTACAAB0IYwAAABdCCMAAEAXwggAANCFMAIAAHQhjAAAAF0IIwAAQBfCCAAA0IUwAgAAdCGMAAAAXQgjAABAF8IIAADQhTACAAB0IYwAAABdCCMAAEAXwggAANCFMAIAAHQhjAAAAF0IIwAAQBfCCAAA0IUwAgAAdCGMAAAAXQgjAABAF8IIAADQhTACAAB0IYwAAABdCCMAAEAXwggAANCFMAIAAHQhjAAAAF0IIwAAQBfCCAAA0MXchZGqOqaq3lRVF1fVdVXVquqcbdQcXlXnV9XVVXVjVV1WVS+oqt22UnN0VV1YVZur6vqq+nhVHbeNfo6rqk+M7TeP9Udvpf1uVXXSOJ4bx/GdX1WHb/s3AQAA823uwkiSlyU5IcnPJPnGthpX1ROSXJTkiCTvTnJmkjsmeX2Sc5eoOSHJhiQPTXJOkrcmuXeS9VV1+hI1pydZn+SAsf05SR6WZMN4vun2Nfb/unE8Z47jOyLJReO4AQBgzZrHMHJSkoOS3DXJb2+tYVXdNUMwuDXJka21Z7XWXpQhyHwsyTFVdexUzbokpye5OslhrbXnt9ZOSvJTSb6c5OSq+oWpmsOTnDwe/6nW2kmttecnOXQ8z+njeScdm+SYJBuT/Exr7UWttWcledQ43rdW1V2W+0sBAIB5M3dhpLX2odbaF1trbRnNj0lyjyTnttY+NXGOmzJcYUluG2iOT7JHkjNba5smaq5J8sfj0+dN1Sw8f9XYbqFmU5I3j+d75lTNQr8vG8ezUPPJJO8cx33MNn9CAACYU3MXRlbo0eP2/YscuyjJliSHV9Uey6x531Sb7aqpqj2THD72f/EK+gEAgDVj994D2MkeNG6vmD7QWrulqq5M8pAk909y+TJqrqqqG5IcWFV7tda2VNXeSe6T5PrW2lWLjOGL4/agiX0PSLJbkq+01m5ZZs2SquqSJQ49eDn1AADQw1q/MrLPuN28xPGF/ftuR80+U9ud0ce+SxwHAIC5t9avjOwSWmuHLrZ/vGJyyCoPBwAAlmWtXxmZvooxbWH/tdtRs3lquzP6uHaJ4wAAMPfWehj5wri9zdyLqto9yf2S3JLkK8usOSDJ3km+3lrbkiSttRsyrHdy5/H4tAeO28k5KF/OcPve+4/jWE4NAACsKWs9jFwwbh+7yLEjkuyVZGNr7eZl1jxuqs121Yy38t049v9LK+gHAADWjLUeRs5L8n2Rw0EAAAzmSURBVN0kx1bVYQs7x1vrnjY+fctUzdlJbk5ywuRChVW1X5KXjE/PmqpZeP7Ssd1Czbokzx/Pd/ZUzUK/p43jWah5eJKnJPlOkndt4+cDAIC5NXcT2KvqiUmeOD6917j9hapaP/7v77bWXpgkrbXrquo5GULJhVV1boYV0R+f4Ra+52VYYPBHWmtXVtWLkpyR5FNV9c4kP8iwAOGBSV7bWvvYVM3Gqnpdkt9LcllVnZfkjhlCxd2SnDi5gOLo3CRPGs/76arakGT/sWa3JM9prV23Hb8iAACYC3MXRpL8TJLjpvbdf3wkyVeTvHDhQGvtPVX1yCQvTfLkJHsm+VKG4HDGYiu5t9beVFWbxvM8PcMVpM9lWC397YsNqrV2clV9JsOVkOcm+WGSS5O8prX23kXat6p6aoavax2f5MQkN2VYjPG01trGbf8qAABgfs1dGGmtnZLklBXWfDTJr66wZkOSDSusWZ9k/Qra35Lk9eMDAAB2KWt9zggAAHA7JYwAAABdCCMAAEAXwggAANCFMAIAAHQhjAAAAF0IIwAAQBfCCAAA0IUwAgAAdCGMAAAAXQgjAABAF8IIAADQhTACAAB0IYwAAABdCCMAAEAXwggAANCFMAIAAHQhjAAAAF0IIwAAQBfCCAAA0IUwAgAAdCGMAAAAXQgjAABAF8IIAADQhTACAAB0IYwAAABdCCMAAEAXwggAANCFMAIAAHQhjAAAAF0IIwAAQBfCCAAA0IUwAgAAdCGMAAAAXQgjAABAF8IIAADQhTACAAB0IYwAAABdCCMAAEAXwggAANCFMAIAAHQhjAAAAF0IIwAAQBfCCAAA0IUwAgAAdCGMAAAAXQgjAABAF8IIAADQhTACAAB0IYwAAABdCCMAAEAXwggAANCFMAIAAHQhjAAAAF0IIwAAQBfCCAAA0IUwAgAAdCGMAAAAXQgjAABAF8IIAADQhTACAAB0IYwAAABdCCMAAEAXwggAANCFMAIAAHQhjAAAAF0IIwAAQBfCCAAA0IUwAgAAdCGMAAAAXQgjAABAF8IIAADQhTACAAB0IYwAAABdCCOdVdWBVfW2qvpmVd1cVZuq6g1VtV/vsQEAwM60e+8B7Mqq6gFJNia5Z5K/TvL5JP8+ye8meWxVPaK19r2OQwQAgJ3GlZG+/iRDEPmd1toTW2t/0Fp7dJLXJ3lQkld1HR0AAOxEwkgn41WRxyTZlOTNU4dfkeSGJE+rqr1XeWgAALAqhJF+HjVuP9Ba++Hkgdba95N8NMleSX5+tQcGAACrwZyRfh40bq9Y4vgXM1w5OSjJB7d2oqq6ZIlDP3355Zfn0EMP3b4R7oCrvrF51fvsbY/1v9t7CKvuoffZp/cQAIDOLr/88iRZtz21wkg/C3/FLfVX+8L+fXegj1tvvPHGzZdeeummHTjHSj143H5+Ffukk0u/vehu7wES7wMG3gck3ge7gnVJrtueQmFkDWitrf6ljyUsXKW5PY2J1eU9QOJ9wMD7gMT7gK0zZ6SfhSsfS33PZWH/taswFgAAWHXCSD9fGLcHLXH8geN2qTklAAAw14SRfj40bh9TVf/mdaiquyR5RJItSf5htQcGAACrQRjppLX25SQfyDDh5/lTh09NsneSd7TWbljloQEAwKowgb2v/5JkY5IzquqoJJcn+bkMa5BckeSlHccGAAA7VbXWeo9hl1ZV903yyiSPTbJ/kquSvDvJqa21a3qODQAAdiZhBAAA6MKcEQAAoAthBAAA6EIYAQAAuhBGAACALoQRAACgC2EEAADoQhhhh1XVgVX1tqr6ZlXdXFWbquoNVbVf77ExW+Nr25Z4fGuJmsOr6vyqurqqbqyqy6rqBVW122qPn+WrqmOq6k1VdXFVXTe+xudso2bFr3VVHV1VF1bV5qq6vqo+XlXHzf4nYnus5H1QVeu28vnQqurcrfRzXFV9YnwPbB7fE0fvvJ+M5aqq/avq2VX17qr60vj/7c1V9ZGqelZVLfq3pM8DlssK7OyQqnpAhlXk75nkr5N8Psm/T/K7SR5bVY9orX2v4xCZvc1J3rDI/uund1TVE5K8K8lNSd6Z5Ookv5bk9UkekeQ3dt4w2UEvS/LTGV7Xryd58NYab89rXVUnJHlTku8lOSfJD5Ick2R9VT2stfbCWf0wbLcVvQ9G/zvJexbZ/9nFGlfV6UlOHs//1iR3THJskg1VdWJr7cztGDez8xtJ3pJhUeYPJflakh9P8qQkf5rkcVX1G21i4TqfB6xIa83DY7sfSf42SUty4tT+1437z+o9Ro+Zvt6bkmxaZtu7JvnnJDcnOWxi/54ZAmxLcmzvn8ljydfvUUkemKSSHDm+XufM6rVOsi7DHyrfS7JuYv9+Sb401vxC79/Drv5Y4ftg3Xh8/QrOf/hY86Uk+02d63vje2TdjvwMHjv8Hnh0hiBxh6n998oQTFqSJ0/s93ngsaKHr2mx3carIo/J8Afqm6cOvyLJDUmeVlV7r/LQuH04Jsk9kpzbWvvUws7W2k0Z/rU1SX67x8DYttbah1prX2zjXwTbsD2v9fFJ9khyZmtt00TNNUn+eHz6vO0cPjOywvfB9lh4jV81vvYL/W7K8N+VPZI8cyf1zTK01i5orW1orf1wav+3kpw1Pj1y4pDPA1ZEGGFHPGrcfmCRD6nvJ/lokr2S/PxqD4ydao+q+s2qeklV/W5VPWqJ7wA/ety+f5FjFyXZkuTwqtpjp42U1bI9r/XWat431Yb5cu+q+q3xM+K3quqnttLW+2C+/cu4vWVin88DVsScEXbEg8btFUsc/2KGKycHJfngqoyI1XCvJO+Y2ndlVT2ztfbhiX1Lvj9aa7dU1ZVJHpLk/kku3ykjZbVsz2u9tZqrquqGJAdW1V6ttS07YczsPL88Pn6kqi5Mclxr7WsT+/ZOcp8k17fWrlrkPF8ctwftpHGyA6pq9yRPH59OhgifB6yIKyPsiH3G7eYlji/s33cVxsLqODvJURkCyd5JHpbkv2f4vu/7quqnJ9p6f+w6tue1Xm7NPksc5/ZnS5I/SnJohu/675fkkRkmPR+Z5INTX9v1GTHfXp3koUnOb6397cR+nwesiDACLFtr7dTx+8Pfbq1taa19trX2vAw3LLhTklP6jhDopbX2z621P2ytXdpau3Z8XJThCvnHk/xkkmf3HSWzUFW/k+EOaJ9P8rTOw2HOCSPsiG39S8XC/mtXYSz0tTCJ8YiJfd4fu47tea2XW7PUv5QyJ1prt2S4BWziM2LujbfgfWOSzyV5VGvt6qkmPg9YEWGEHfGFcbvU93kfOG6XmlPC2vGdcTv5FYwl3x/jd43vl2HS41d27tBYBdvzWm+t5oAM76Wv+374mnGbz4jW2g1JvpHkzuNrPs1/Q25nquoFGdYC+WyGILLYYrc+D1gRYYQd8aFx+5jpFVir6i4ZFjbakuQfVntgrLqFO6ZN/sflgnH72EXaH5HhTmsbW2s378yBsSq257XeWs3jptow/xb7jEi8D+ZGVf1+hkUL/zFDEPnnJZr6PGBFhBG2W2vty0k+kGHy8vOnDp+a4V8y3jH+6xdzrqoOXmzNmKpal2RhheRzJg6dl+S7SY6tqsMm2u+Z5LTx6Vt2ymBZbdvzWp+dYVG0E8b30ELNfkleMj49K8yNqjpk+h+mxv1HJTlpfHrO1OGF1/il42u/ULMuw39Xbs7wXqGjqnp5hgnrlyQ5qrX23a0093nAitTOW8eIXcG48OHGJPdM8tcZbtP3cxnWILkiyeGtte/1GyGzUlWnZJiweFGSryb5fpIHJPkPGVbWPT/Jr7fWfjBR88QM/2G6Kcm5Sa5O8vgMt3E8L8l/3ImLqbEDxtfuiePTeyX5lQz/qn3xuO+7rbUXTrVf0WtdVScmOSPDqsvvTPKDDAumHZjktZPnp4+VvA/G2/c+MMN/E74+Hv+p/Ov6EC9vrS38MTrZx2uT/N5Yc16SOyZ5SpL9k5zYWjtzuobVU1XHJVmf5NYMX9FabN7Gptba+okanwcsmzDCDquq+yZ5ZYbLq/snuSrJu5OcOrmiLvOtqh6ZYQXcn82/3tr32gyX7N+R4SrYbT5QquoRSV6a5BcyhJYvJXlbkjNaa7euzuhZqTF8vmIrTb7aWls3VbPi17qqfi3JC5MckuFq/ecyrML89h38EZiBlbwPqupZSX49w+1e757kx5J8O8nHMrymFy91kqp6RoYrIf8uyQ+TXJrkNa219+7wD8EOWcZ7IEk+3Fo7cqrO5wHLIowAAABdmDMCAAB0IYwAAABdCCMAAEAXwggAANCFMAIAAHQhjAAAAF0IIwAAQBfCCAAA0IUwAgAAdCGMAAAAXQgjAABAF8IIAADQhTACAAB0IYwAAABdCCMAAEAXwggAANCFMAIAAHTxfwCYUkdVMjWCsgAAAABJRU5ErkJggg==\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "image/png": {
       "height": 248,
       "width": 401
      },
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "loci_df.MEAN_DEPTH.plot(kind=\"hist\");"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0"
      ]
     },
     "execution_count": 63,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(loci_df[loci_df.Fis == -9])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(46363, 46491)"
      ]
     },
     "execution_count": 64,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(loci_df[loci_df.SUM_DEPTH <5000]),len(loci_df[loci_df.SUM_DEPTH >=500])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 152,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(21, 17278)"
      ]
     },
     "execution_count": 152,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(loci_df[loci_df.MEAN_DEPTH > 25 ]),len(loci_df[loci_df.MEAN_DEPTH >= 2])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 143,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0"
      ]
     },
     "execution_count": 143,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(loci_df[loci_df.QUAL >= 20]) - len(loci_df[loci_df.QUAL >= 30])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 144,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(46854, 46848, 46799)"
      ]
     },
     "execution_count": 144,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(loci_df[loci_df.QUAL >  20]), len(loci_df[loci_df.QUAL > 50]), len(loci_df[loci_df.QUAL > 100])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 145,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(46726, 0)"
      ]
     },
     "execution_count": 145,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    " len(loci_df[loci_df.Fis >= -0.5]), len(loci_df[loci_df.MAF < 0.01])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 153,
   "metadata": {},
   "outputs": [],
   "source": [
    "def filter_snps(df, imputed=False):\n",
    "    if imputed:\n",
    "        return df[(df.MAF >= 0.01) &  \n",
    "                  (df.Fis > -0.5)]\n",
    "    else:\n",
    "        return df[(df.MEAN_DEPTH >= 2) & \n",
    "                  (df.MEAN_DEPTH < 25) & \n",
    "                  (df.QUAL >= 100) & \n",
    "                  (df.MAF >= 0.01) &\n",
    "                  (df.Fis >= -0.5)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 154,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(17201, 25)"
      ]
     },
     "execution_count": 154,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#42513 with df.Fis >= -0.5\n",
    "#42688 without \n",
    "loci_stage1 = filter_snps(loci_df)\n",
    "loci_stage1.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 155,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(os.path.join(analysis_dir, \"stage1_positions.txt\"), \"w\") as o:\n",
    "    for elem in loci_stage1.index:\n",
    "        o.write(\"%s\\n\" % \"\\t\".join(elem.split(\"-\")))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 156,
   "metadata": {
    "run_control": {
     "marked": false
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "VCFtools - 0.1.16\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf /data/gpfs/assoc/denovo/tfaske/sagebrush/SNPcall/filtering/snps_indv_removed.recode.vcf.gz\n",
      "\t--recode-INFO-all\n",
      "\t--out /data/gpfs/assoc/denovo/tfaske/sagebrush/SNPcall/filtering/good_snps\n",
      "\t--positions /data/gpfs/assoc/denovo/tfaske/sagebrush/SNPcall/filtering/stage1_positions.txt\n",
      "\t--recode\n",
      "\t--remove-filtered-all\n",
      "\t--remove-indels\n",
      "\n",
      "Using zlib version: 1.2.11\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=PV4,Number=4,Type=Float,Description=\"P-values for strand bias, baseQ bias, mapQ bias and tail distance bias\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=PV4,Number=4,Type=Float,Description=\"P-values for strand bias, baseQ bias, mapQ bias and tail distance bias\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=DP4,Number=4,Type=Integer,Description=\"Number of high-quality ref-forward , ref-reverse, alt-forward and alt-reverse bases\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=DP4,Number=4,Type=Integer,Description=\"Number of high-quality ref-forward , ref-reverse, alt-forward and alt-reverse bases\">\n",
      "After filtering, kept 698 out of 698 Individuals\n",
      "Outputting VCF file...\n",
      "After filtering, kept 17201 out of a possible 46854 Sites\n",
      "Run Time = 25.00 seconds\n"
     ]
    }
   ],
   "source": [
    "!$vcftools --gzvcf $vcf_bad_remove_gz \\\n",
    "--remove-indels  \\\n",
    "--remove-filtered-all \\\n",
    "--recode \\\n",
    "--recode-INFO-all \\\n",
    "--positions {os.path.join(analysis_dir, \"stage1_positions.txt\")} \\\n",
    "--out {os.path.join(analysis_dir, \"good_snps\")}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 157,
   "metadata": {},
   "outputs": [],
   "source": [
    "snps = os.path.join(analysis_dir, \"good_snps.recode.vcf\")\n",
    "snps_gz = snps + \".gz\"\n",
    "!$bgzip -c {snps} > {snps_gz}\n",
    "!$tabix {snps_gz}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## recalculate missing data (loci / indv) & remove alleles not present in one ssp_ploidy group"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 158,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R\n",
    "vcf <- read.vcfR(\"good_snps.recode.vcf\", verbose = FALSE)\n",
    "\n",
    "#get positions\n",
    "chrom <- getCHROM(vcf)\n",
    "pos <- getPOS(vcf)\n",
    "pos_ID <- paste(chrom,pos,sep = ':')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 159,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R\n",
    "#get DP \n",
    "dp <- extract.gt(vcf, element = 'DP')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 160,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] 17201\n",
      " [1] \"CM042338.1:1496841\" \"CM042338.1:1942811\" \"CM042338.1:6353755\"\n",
      " [4] \"CM042338.1:7214312\" \"CM042338.1:8063221\" \"CM042338.1:8177835\"\n",
      " [7] \"CM042338.1:8530959\" \"CM042338.1:8548932\" \"CM042338.1:8655617\"\n",
      "[10] \"CM042338.1:8690402\"\n",
      "                   AT2_BB_10 AT2_BB_11 AT2_BB_12 AT2_BB_13 AT2_BB_14\n",
      "CM042338.1_1496841 \"0\"       \"1\"       \"3\"       \"6\"       \"8\"      \n",
      "CM042338.1_1942811 \"0\"       \"0\"       \"0\"       \"0\"       \"9\"      \n",
      "CM042338.1_6353755 \"7\"       \"3\"       \"5\"       \"5\"       \"6\"      \n",
      "CM042338.1_7214312 \"5\"       \"3\"       \"0\"       \"0\"       \"0\"      \n",
      "CM042338.1_8063221 \"3\"       \"3\"       \"0\"       \"1\"       \"8\"      \n"
     ]
    }
   ],
   "source": [
    "%%R\n",
    "## check out PL and pos_ID\n",
    "print(length(pos_ID))\n",
    "print(pos_ID[1:10])\n",
    "dp[1:5,1:5]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 161,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] 698\n",
      " [1] \"AT2_BB_10\" \"AT2_BB_11\" \"AT2_BB_12\" \"AT2_BB_13\" \"AT2_BB_14\" \"AT2_BB_15\"\n",
      " [7] \"AT2_BB_16\" \"AT2_BB_17\" \"AT2_BB_1\"  \"AT2_BB_2\" \n",
      " [1] \"AT2\" \"AT2\" \"AT2\" \"AT2\" \"AT2\" \"AT2\" \"AT2\" \"AT2\" \"AT2\" \"AT2\"\n",
      "AT2 AT2 AT2 AT2 AT2 AT2 AT2 AT2 AT2 AT2 \n",
      "\"2\" \"2\" \"2\" \"2\" \"2\" \"2\" \"2\" \"2\" \"2\" \"2\" \n"
     ]
    }
   ],
   "source": [
    "%%R\n",
    "ID <- colnames(dp)\n",
    "ssp_ploid <- as.character(sapply(ID,function(s) unlist(strsplit(as.character(s),'_'))[1]))\n",
    "ploidy <- sapply(ssp_ploid,function(s) gsub('(\\\\D)','',s,perl=TRUE))\n",
    "print(length(ID))\n",
    "print(ID[1:10])\n",
    "print(ssp_ploid[1:10])\n",
    "print(ploidy[1:10])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Calculate missing loci and individual"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 162,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] 17201\n",
      "[1] 698\n"
     ]
    }
   ],
   "source": [
    "%%R\n",
    "nindv <- ncol(dp)\n",
    "nloci <- nrow(dp)\n",
    "print(nloci)\n",
    "print(nindv)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 163,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] 17201\n",
      "    Min.  1st Qu.   Median     Mean  3rd Qu.     Max. \n",
      "0.001433 0.128940 0.196275 0.207378 0.270774 0.590258 \n"
     ]
    }
   ],
   "source": [
    "%%R\n",
    "miss_loci <- apply(dp,1, function(d) length(which(d == \"0\"))/nindv)\n",
    "#miss_loci <- apply(dp[,1:100],1, function(d) (length(which(d == \"0\"))/100))\n",
    "print(length(miss_loci))\n",
    "print(summary(miss_loci))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 168,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "   Min. 1st Qu.  Median    Mean 3rd Qu.    Max. \n",
      "0.00000 0.04298 0.06447 0.07052 0.08883 0.29943 \n",
      "    Min.  1st Qu.   Median     Mean  3rd Qu.     Max. \n",
      "0.000000 0.002865 0.005731 0.006708 0.008596 0.037249 \n",
      "   Min. 1st Qu.  Median    Mean 3rd Qu.    Max. \n",
      "0.00000 0.03582 0.05587 0.06533 0.08023 0.24212 \n",
      "   Min. 1st Qu.  Median    Mean 3rd Qu.    Max. \n",
      "0.00000 0.01146 0.01862 0.02178 0.02865 0.10602 \n",
      "   Min. 1st Qu.  Median    Mean 3rd Qu.    Max. \n",
      "0.00000 0.01862 0.03438 0.04303 0.05731 0.27364 \n"
     ]
    }
   ],
   "source": [
    "%%R\n",
    "#calculaate missing within each spp_ploid (hard code it)\n",
    "\n",
    "#subset dp\n",
    "dp_AT2 <- dp[,which(ssp_ploid=='AT2')]\n",
    "dp_AT4 <- dp[,which(ssp_ploid=='AT4')]\n",
    "dp_AV2 <- dp[,which(ssp_ploid=='AV2')]\n",
    "dp_AV4 <- dp[,which(ssp_ploid=='AV4')]\n",
    "dp_AW4 <- dp[,which(ssp_ploid=='AW4')]\n",
    "\n",
    "#calc miss\n",
    "miss_loci_AT2 <- apply(dp_AT2,1, function(d) length(which(d == \"0\"))/nindv)\n",
    "print(summary(miss_loci_AT2))\n",
    "                       \n",
    "miss_loci_AT4 <- apply(dp_AT4,1, function(d) length(which(d == \"0\"))/nindv)\n",
    "print(summary(miss_loci_AT4))\n",
    "                       \n",
    "miss_loci_AV2 <- apply(dp_AV2,1, function(d) length(which(d == \"0\"))/nindv)\n",
    "print(summary(miss_loci_AV2))                       \n",
    "                       \n",
    "miss_loci_AV4 <- apply(dp_AV4,1, function(d) length(which(d == \"0\"))/nindv)\n",
    "print(summary(miss_loci_AV4))                       \n",
    "                       \n",
    "miss_loci_AW4 <- apply(dp_AW4,1, function(d) length(which(d == \"0\"))/nindv)\n",
    "print(summary(miss_loci_AW4))                       "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 169,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "    Min.  1st Qu.   Median     Mean  3rd Qu.     Max. \n",
      "0.000000 0.003488 0.005232 0.005513 0.007093 0.021278 \n",
      "    Min.  1st Qu.   Median     Mean  3rd Qu.     Max. \n",
      "0.000000 0.001453 0.002500 0.002902 0.003779 0.016743 \n"
     ]
    }
   ],
   "source": [
    "%%R\n",
    "#calculaate missing within each ploidy\n",
    "\n",
    "#subset dp\n",
    "dp2 <- dp[,which(ploidy=='2')]\n",
    "dp4 <- dp[,which(ploidy=='4')]\n",
    "\n",
    "#calc miss\n",
    "miss_loci_2 <- apply(dp2,1, function(d) length(which(d == \"0\"))/nloci)\n",
    "print(summary(miss_loci_2))\n",
    "                       \n",
    "miss_loci_4 <- apply(dp4,1, function(d) length(which(d == \"0\"))/nloci)\n",
    "print(summary(miss_loci_4))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### all groups have v little missing data, remove none"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Miss indv!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 170,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] 698\n",
      "   Min. 1st Qu.  Median    Mean 3rd Qu.    Max. \n",
      " 0.0611  0.1358  0.1853  0.2074  0.2507  0.8839 \n"
     ]
    }
   ],
   "source": [
    "%%R\n",
    "nloci <- nrow(dp)\n",
    "#print(nloci)\n",
    "\n",
    "miss_indv <- apply(dp,2, function(d) length(which(d == \"0\"))/nloci)\n",
    "print(length(miss_indv))\n",
    "print(summary(miss_indv))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 171,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"number of individuals with greater than 10% missing data.....  656\"\n",
      "[1] \"number of individuals with greater than 20% missing data.....  314\"\n",
      "[1] \"number of individuals with greater than 30% missing data.....  99\"\n",
      "[1] \"number of individuals with greater than 40% missing data.....  27\"\n",
      "[1] \"number of individuals with greater than 50% missing data.....  12\"\n",
      "[1] \"number of individuals with greater than 60% missing data.....  8\"\n",
      "[1] \"number of individuals with greater than 70% missing data.....  6\"\n",
      "[1] \"number of individuals with greater than 80% missing data.....  2\"\n",
      "[1] \"number of individuals with greater than 90% missing data.....  0\"\n"
     ]
    }
   ],
   "source": [
    "%%R\n",
    "for (perc in seq(.1,.9,by=.1)){\n",
    "    print(paste0('number of individuals with greater than ',perc*100,'% missing data.....  ',\n",
    "                 length(which(miss_indv >= perc))))\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 172,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R\n",
    "miss60_indv_df <- data.frame(INDV=ID[which(miss_indv >= .60)])\n",
    "miss60_indv_df\n",
    "write.table(miss60_indv_df,'miss60_indv.txt',quote=F,row.names=F,col.names=T)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### use vcftools to remove individuals and loci based off above parameters"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 173,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "VCFtools - 0.1.16\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf /data/gpfs/assoc/denovo/tfaske/sagebrush/SNPcall/filtering/good_snps.recode.vcf.gz\n",
      "\t--remove /data/gpfs/assoc/denovo/tfaske/sagebrush/SNPcall/filtering/miss60_indv.txt\n",
      "\t--recode-INFO-all\n",
      "\t--out /data/gpfs/assoc/denovo/tfaske/sagebrush/SNPcall/filtering/final_snps\n",
      "\t--recode\n",
      "\t--remove-filtered-all\n",
      "\n",
      "Using zlib version: 1.2.11\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=PV4,Number=4,Type=Float,Description=\"P-values for strand bias, baseQ bias, mapQ bias and tail distance bias\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=PV4,Number=4,Type=Float,Description=\"P-values for strand bias, baseQ bias, mapQ bias and tail distance bias\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=DP4,Number=4,Type=Integer,Description=\"Number of high-quality ref-forward , ref-reverse, alt-forward and alt-reverse bases\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=DP4,Number=4,Type=Integer,Description=\"Number of high-quality ref-forward , ref-reverse, alt-forward and alt-reverse bases\">\n",
      "Excluding individuals in 'exclude' list\n",
      "After filtering, kept 690 out of 698 Individuals\n",
      "Outputting VCF file...\n",
      "After filtering, kept 17201 out of a possible 17201 Sites\n",
      "Run Time = 23.00 seconds\n"
     ]
    }
   ],
   "source": [
    "!$vcftools --gzvcf $snps_gz \\\n",
    "--remove-filtered-all \\\n",
    "--recode \\\n",
    "--recode-INFO-all \\\n",
    "--remove {os.path.join(analysis_dir, \"miss60_indv.txt\")} \\\n",
    "--out {os.path.join(analysis_dir, \"final_snps\")}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 174,
   "metadata": {},
   "outputs": [],
   "source": [
    "final_snps = os.path.join(analysis_dir, \"final_snps.recode.vcf\")\n",
    "final_snps_gz = final_snps + \".gz\"\n",
    "!$bgzip -c {final_snps} > {final_snps_gz}\n",
    "!$tabix {final_snps_gz}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Make 012"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 175,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "VCFtools - 0.1.16\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf /data/gpfs/assoc/denovo/tfaske/sagebrush/SNPcall/filtering/final_snps.recode.vcf\n",
      "\t--012\n",
      "\t--out /data/gpfs/assoc/denovo/tfaske/sagebrush/SNPcall/filtering/final_snps.recode.vcf\n",
      "\n",
      "Using zlib version: 1.2.11\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=PV4,Number=4,Type=Float,Description=\"P-values for strand bias, baseQ bias, mapQ bias and tail distance bias\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=PV4,Number=4,Type=Float,Description=\"P-values for strand bias, baseQ bias, mapQ bias and tail distance bias\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=DP4,Number=4,Type=Integer,Description=\"Number of high-quality ref-forward , ref-reverse, alt-forward and alt-reverse bases\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=DP4,Number=4,Type=Integer,Description=\"Number of high-quality ref-forward , ref-reverse, alt-forward and alt-reverse bases\">\n",
      "After filtering, kept 690 out of 690 Individuals\n",
      "Writing 012 matrix files ... Done.\n",
      "After filtering, kept 17201 out of a possible 17201 Sites\n",
      "Run Time = 3.00 seconds\n"
     ]
    }
   ],
   "source": [
    "f = os.path.join(analysis_dir, \"final_snps.recode.vcf\")\n",
    "!$vcftools --gzvcf {f} \\\n",
    "--out {f} \\\n",
    "--012"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Get coverage per individual"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 176,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "VCFtools - 0.1.16\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf /data/gpfs/assoc/denovo/tfaske/sagebrush/SNPcall/filtering/final_snps.recode.vcf\n",
      "\t--depth\n",
      "\t--out /data/gpfs/assoc/denovo/tfaske/sagebrush/SNPcall/filtering/final_snps.recode.vcf\n",
      "\n",
      "Using zlib version: 1.2.11\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=PV4,Number=4,Type=Float,Description=\"P-values for strand bias, baseQ bias, mapQ bias and tail distance bias\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=PV4,Number=4,Type=Float,Description=\"P-values for strand bias, baseQ bias, mapQ bias and tail distance bias\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=DP4,Number=4,Type=Integer,Description=\"Number of high-quality ref-forward , ref-reverse, alt-forward and alt-reverse bases\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=DP4,Number=4,Type=Integer,Description=\"Number of high-quality ref-forward , ref-reverse, alt-forward and alt-reverse bases\">\n",
      "After filtering, kept 690 out of 690 Individuals\n",
      "Outputting Mean Depth by Individual\n",
      "After filtering, kept 17201 out of a possible 17201 Sites\n",
      "Run Time = 2.00 seconds\n"
     ]
    }
   ],
   "source": [
    "f = os.path.join(analysis_dir, \"final_snps.recode.vcf\")\n",
    "!$vcftools --gzvcf {f} \\\n",
    "--out {f} \\\n",
    "--depth"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 177,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>INDV</th>\n",
       "      <th>N_SITES</th>\n",
       "      <th>MEAN_DEPTH</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AT2_BB_10</td>\n",
       "      <td>17201</td>\n",
       "      <td>4.41079</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AT2_BB_11</td>\n",
       "      <td>17201</td>\n",
       "      <td>2.74972</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AT2_BB_12</td>\n",
       "      <td>17201</td>\n",
       "      <td>2.36021</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AT2_BB_13</td>\n",
       "      <td>17201</td>\n",
       "      <td>3.42829</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AT2_BB_14</td>\n",
       "      <td>17201</td>\n",
       "      <td>4.72223</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        INDV  N_SITES  MEAN_DEPTH\n",
       "0  AT2_BB_10    17201     4.41079\n",
       "1  AT2_BB_11    17201     2.74972\n",
       "2  AT2_BB_12    17201     2.36021\n",
       "3  AT2_BB_13    17201     3.42829\n",
       "4  AT2_BB_14    17201     4.72223"
      ]
     },
     "execution_count": 177,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "depth_file = os.path.join(analysis_dir, \"final_snps.recode.vcf.idepth\")\n",
    "depth_df = pd.read_csv(depth_file, sep=\"\\t\")\n",
    "depth_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 178,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "count    690.000000\n",
       "mean       3.335573\n",
       "std        1.152326\n",
       "min        0.914656\n",
       "25%        2.467112\n",
       "50%        3.198155\n",
       "75%        4.126478\n",
       "max        7.355390\n",
       "Name: MEAN_DEPTH, dtype: float64"
      ]
     },
     "execution_count": 178,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "depth_df.MEAN_DEPTH.describe()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Remove unnecessary files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 179,
   "metadata": {},
   "outputs": [],
   "source": [
    "!rm *vcf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 1
}
